diff --git a/arch/x86/include/asm/suspend_64.h b/arch/x86/include/asm/suspend_64.h
index 6136a18..585dcbd 100644
--- a/arch/x86/include/asm/suspend_64.h
+++ b/arch/x86/include/asm/suspend_64.h
@@ -44,6 +44,6 @@ struct saved_context {
 /* routines for saving/restoring kernel state */
 extern int acpi_save_state_mem(void);
 extern char core_restore_code;
-extern char restore_registers;
+extern int restore_registers(void);
 
 #endif /* _ASM_X86_SUSPEND_64_H */
