diff --git a/arch/x86/include/asm/page_32.h b/arch/x86/include/asm/page_32.h
index 904f528..b4d0d24 100644
--- a/arch/x86/include/asm/page_32.h
+++ b/arch/x86/include/asm/page_32.h
@@ -7,11 +7,17 @@
 
 #define __phys_addr_nodebug(x)	((x) - PAGE_OFFSET)
 #ifdef CONFIG_DEBUG_VIRTUAL
-extern unsigned long __phys_addr(unsigned long);
+extern unsigned long __intentional_overflow(-1) __phys_addr(unsigned long);
 #else
-#define __phys_addr(x)		__phys_addr_nodebug(x)
+static inline unsigned long __intentional_overflow(-1) __phys_addr(unsigned long x)
+{
+	return __phys_addr_nodebug(x);
+}
 #endif
-#define __phys_addr_symbol(x)	__phys_addr(x)
+static inline unsigned long __intentional_overflow(-1) __phys_addr_symbol(unsigned long x)
+{
+	return __phys_addr(x);
+}
 #define __phys_reloc_hide(x)	RELOC_HIDE((x), 0)
 
 #ifdef CONFIG_FLATMEM
