diff --git a/arch/x86/math-emu/wm_shrx.S b/arch/x86/math-emu/wm_shrx.S
index 5184283..6267001 100644
--- a/arch/x86/math-emu/wm_shrx.S
+++ b/arch/x86/math-emu/wm_shrx.S
@@ -54,7 +54,7 @@ ENTRY(FPU_shrx)
 	popl	%ebx
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_shrx
 
 L_more_than_31:
 	cmpl	$64,%ecx
@@ -69,7 +69,7 @@ L_more_than_31:
 	movl	$0,4(%esi)
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_shrx
 
 L_more_than_63:
 	cmpl	$96,%ecx
@@ -83,7 +83,7 @@ L_more_than_63:
 	movl	%edx,4(%esi)
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_shrx
 
 L_more_than_95:
 	xorl	%eax,%eax
@@ -91,7 +91,7 @@ L_more_than_95:
 	movl	%eax,4(%esi)
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_shrx
 
 
 /*---------------------------------------------------------------------------+
@@ -144,7 +144,7 @@ ENTRY(FPU_shrxs)
 	popl	%ebx
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_shrxs
 
 /* Shift by [0..31] bits */
 Ls_less_than_32:
@@ -161,7 +161,7 @@ Ls_less_than_32:
 	popl	%ebx
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_shrxs
 
 /* Shift by [64..95] bits */
 Ls_more_than_63:
@@ -187,7 +187,7 @@ Ls_more_than_63:
 	popl	%ebx
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_shrxs
 
 Ls_more_than_95:
 /* Shift by [96..inf) bits */
@@ -201,4 +201,4 @@ Ls_more_than_95:
 	popl	%ebx
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_shrxs
