diff --git a/arch/mips/kernel/process.c b/arch/mips/kernel/process.c
index fbbf5fc..eec61dd 100644
--- a/arch/mips/kernel/process.c
+++ b/arch/mips/kernel/process.c
@@ -622,18 +622,6 @@ unsigned long get_wchan(struct task_struct *task)
 	return pc;
 }
 
-/*
- * Don't forget that the stack pointer must be aligned on a 8 bytes
- * boundary for 32-bits ABI and 16 bytes for 64-bits ABI.
- */
-unsigned long arch_align_stack(unsigned long sp)
-{
-	if (!(current->personality & ADDR_NO_RANDOMIZE) && randomize_va_space)
-		sp -= get_random_int() & ~PAGE_MASK;
-
-	return sp & ALMASK;
-}
-
 static DEFINE_PER_CPU(struct call_single_data, backtrace_csd);
 static struct cpumask backtrace_csd_busy;
 
