diff --git a/arch/x86/xen/xen-head.S b/arch/x86/xen/xen-head.S
index 7f8d8ab..3032b77 100644
--- a/arch/x86/xen/xen-head.S
+++ b/arch/x86/xen/xen-head.S
@@ -8,6 +8,7 @@
 
 #include <asm/boot.h>
 #include <asm/asm.h>
+#include <asm/alternative-asm.h>
 #include <asm/page_types.h>
 
 #include <xen/interface/elfnote.h>
@@ -50,6 +51,18 @@ ENTRY(startup_xen)
 	mov %_ASM_SI, xen_start_info
 	mov $init_thread_union+THREAD_SIZE, %_ASM_SP
 
+#if defined(CONFIG_X86_32) && defined(CONFIG_SMP)
+	movl $cpu_gdt_table,%edi
+	movl $__per_cpu_load,%eax
+	movw %ax,__KERNEL_PERCPU + 2(%edi)
+	rorl $16,%eax
+	movb %al,__KERNEL_PERCPU + 4(%edi)
+	movb %ah,__KERNEL_PERCPU + 7(%edi)
+	movl $__per_cpu_end - 1,%eax
+	subl $__per_cpu_start,%eax
+	movw %ax,__KERNEL_PERCPU + 0(%edi)
+#endif
+
 	jmp xen_start_kernel
 
 	__FINIT
@@ -85,7 +98,7 @@ ENTRY(xen_pvh_early_cpu_init)
 	cmp     $0, %r11b
 	jne     cpu_bringup_and_idle
 #endif
-	ret
+	pax_ret xen_pvh_early_cpu_init
 
 #endif /* CONFIG_XEN_PVH */
 
