diff --git a/arch/x86/lib/memcpy_64.S b/arch/x86/lib/memcpy_64.S
index 9a53a06..275443c 100644
--- a/arch/x86/lib/memcpy_64.S
+++ b/arch/x86/lib/memcpy_64.S
@@ -38,7 +38,7 @@ ENTRY(memcpy)
 	rep movsq
 	movl %edx, %ecx
 	rep movsb
-	ret
+	pax_ret memcpy
 ENDPROC(memcpy)
 ENDPROC(__memcpy)
 EXPORT_SYMBOL(memcpy)
@@ -52,7 +52,7 @@ ENTRY(memcpy_erms)
 	movq %rdi, %rax
 	movq %rdx, %rcx
 	rep movsb
-	ret
+	pax_ret memcpy
 ENDPROC(memcpy_erms)
 
 ENTRY(memcpy_orig)
@@ -136,7 +136,7 @@ ENTRY(memcpy_orig)
 	movq %r9,	1*8(%rdi)
 	movq %r10,	-2*8(%rdi, %rdx)
 	movq %r11,	-1*8(%rdi, %rdx)
-	retq
+	pax_ret memcpy
 	.p2align 4
 .Lless_16bytes:
 	cmpl $8,	%edx
@@ -148,7 +148,7 @@ ENTRY(memcpy_orig)
 	movq -1*8(%rsi, %rdx),	%r9
 	movq %r8,	0*8(%rdi)
 	movq %r9,	-1*8(%rdi, %rdx)
-	retq
+	pax_ret memcpy
 	.p2align 4
 .Lless_8bytes:
 	cmpl $4,	%edx
@@ -161,7 +161,7 @@ ENTRY(memcpy_orig)
 	movl -4(%rsi, %rdx), %r8d
 	movl %ecx, (%rdi)
 	movl %r8d, -4(%rdi, %rdx)
-	retq
+	pax_ret memcpy
 	.p2align 4
 .Lless_3bytes:
 	subl $1, %edx
@@ -179,7 +179,7 @@ ENTRY(memcpy_orig)
 	movb %cl, (%rdi)
 
 .Lend:
-	retq
+	pax_ret memcpy
 ENDPROC(memcpy_orig)
 
 #ifndef CONFIG_UML
@@ -275,7 +275,7 @@ ENTRY(memcpy_mcsafe_unrolled)
 	/* Copy successful. Return zero */
 .L_done_memcpy_trap:
 	xorq %rax, %rax
-	ret
+	pax_ret memcpy
 ENDPROC(memcpy_mcsafe_unrolled)
 EXPORT_SYMBOL_GPL(memcpy_mcsafe_unrolled)
 
@@ -283,7 +283,7 @@ EXPORT_SYMBOL_GPL(memcpy_mcsafe_unrolled)
 	/* Return -EFAULT for any failure */
 .L_memcpy_mcsafe_fail:
 	mov	$-EFAULT, %rax
-	ret
+	pax_ret memcpy
 
 	.previous
 
