diff --git a/arch/x86/xen/irq.c b/arch/x86/xen/irq.c
index 33e9295..7af61ba 100644
--- a/arch/x86/xen/irq.c
+++ b/arch/x86/xen/irq.c
@@ -116,10 +116,10 @@ static void xen_halt(void)
 }
 
 static const struct pv_irq_ops xen_irq_ops __initconst = {
-	.save_fl = PV_CALLEE_SAVE(xen_save_fl),
-	.restore_fl = PV_CALLEE_SAVE(xen_restore_fl),
-	.irq_disable = PV_CALLEE_SAVE(xen_irq_disable),
-	.irq_enable = PV_CALLEE_SAVE(xen_irq_enable),
+	.save_fl = PV_CALLEE_SAVE(save_fl, xen_save_fl),
+	.restore_fl = PV_CALLEE_SAVE(restore_fl, xen_restore_fl),
+	.irq_disable = PV_CALLEE_SAVE(irq_disable, xen_irq_disable),
+	.irq_enable = PV_CALLEE_SAVE(irq_enable, xen_irq_enable),
 
 	.safe_halt = xen_safe_halt,
 	.halt = xen_halt,
