diff --git a/arch/x86/math-emu/reg_round.S b/arch/x86/math-emu/reg_round.S
index d1d4e48..003208e 100644
--- a/arch/x86/math-emu/reg_round.S
+++ b/arch/x86/math-emu/reg_round.S
@@ -436,7 +436,7 @@ fpu_Arith_exit:
 	popl	%edi
 	popl	%esi
 	leave
-	ret
+	pax_ret FPU_round
 
 
 /*
@@ -446,7 +446,7 @@ fpu_Arith_exit:
 L_precision_lost_up:
 	push	%edx
 	push	%eax
-	call	set_precision_flag_up
+	pax_direct_call set_precision_flag_up
 	popl	%eax
 	popl	%edx
 	jmp	L_no_precision_loss
@@ -458,7 +458,7 @@ L_precision_lost_up:
 L_precision_lost_down:
 	push	%edx
 	push	%eax
-	call	set_precision_flag_down
+	pax_direct_call set_precision_flag_down
 	popl	%eax
 	popl	%edx
 	jmp	L_no_precision_loss
@@ -603,7 +603,7 @@ LPseudoDenormal:
 	/* There must be a masked underflow */
 	push	%eax
 	pushl	EX_Underflow
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%eax
 	popl	%eax
 	movl	TAG_Special,%edx
@@ -616,12 +616,12 @@ LPseudoDenormal:
  */
 L_underflow_to_zero:
 	push	%eax
-	call	set_precision_flag_down
+	pax_direct_call set_precision_flag_down
 	popl	%eax
 
 	push	%eax
 	pushl	EX_Underflow
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%eax
 	popl	%eax
 
@@ -635,7 +635,7 @@ L_underflow_to_zero:
 L_overflow:
 	addw	EXTENDED_Ebias,EXP(%edi)	/* Set for unmasked response. */
 	push	%edi
-	call	arith_overflow
+	pax_direct_call arith_overflow
 	pop	%edi
 	jmp	fpu_reg_round_signed_special_exit
 
@@ -653,7 +653,7 @@ Do_unmasked_underflow:
 	addw	$(3*(1<<13)),EXP(%edi)
 	push	%eax
 	pushl	EX_Underflow
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%eax
 	popl	%eax
 	jmp	L_Normalised
@@ -663,44 +663,44 @@ Do_unmasked_underflow:
 #ifdef PECULIAR_486
 L_bugged_denorm_486:
 	pushl	EX_INTERNAL|0x236
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%ebx
 	jmp	L_exception_exit
 #else
 L_bugged_denorm:
 	pushl	EX_INTERNAL|0x230
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%ebx
 	jmp	L_exception_exit
 #endif /* PECULIAR_486 */ 
 
 L_bugged_round24:
 	pushl	EX_INTERNAL|0x231
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%ebx
 	jmp	L_exception_exit
 
 L_bugged_round53:
 	pushl	EX_INTERNAL|0x232
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%ebx
 	jmp	L_exception_exit
 
 L_bugged_round64:
 	pushl	EX_INTERNAL|0x233
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%ebx
 	jmp	L_exception_exit
 
 L_norm_bugged:
 	pushl	EX_INTERNAL|0x234
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%ebx
 	jmp	L_exception_exit
 
 L_entry_bugged:
 	pushl	EX_INTERNAL|0x235
-	call	EXCEPTION
+	pax_direct_call EXCEPTION
 	popl	%ebx
 L_exception_exit:
 	mov	$-1,%eax
