diff --git a/arch/x86/include/asm/mutex_32.h b/arch/x86/include/asm/mutex_32.h
index e9355a8..f031759 100644
--- a/arch/x86/include/asm/mutex_32.h
+++ b/arch/x86/include/asm/mutex_32.h
@@ -30,7 +30,7 @@ do {								\
 								\
 	asm volatile(LOCK_PREFIX "   decl (%%eax)\n"		\
 		     "   jns 1f	\n"				\
-		     "   call " #fail_fn "\n"			\
+		     PAX_DIRECT_CALL(#fail_fn)"\n"		\
 		     "1:\n"					\
 		     : "=a" (dummy)				\
 		     : "a" (count)				\
@@ -76,7 +76,7 @@ do {								\
 								\
 	asm volatile(LOCK_PREFIX "   incl (%%eax)\n"		\
 		     "   jg	1f\n"				\
-		     "   call " #fail_fn "\n"			\
+		     PAX_DIRECT_CALL(#fail_fn)"\n"		\
 		     "1:\n"					\
 		     : "=a" (dummy)				\
 		     : "a" (count)				\
