diff --git a/arch/x86/math-emu/shr_Xsig.S b/arch/x86/math-emu/shr_Xsig.S
index 31cdd11..61a03eca 100644
--- a/arch/x86/math-emu/shr_Xsig.S
+++ b/arch/x86/math-emu/shr_Xsig.S
@@ -44,7 +44,7 @@ ENTRY(shr_Xsig)
 	popl	%ebx
 	popl	%esi
 	leave
-	ret
+	pax_ret shr_Xsig
 
 L_more_than_31:
 	cmpl	$64,%ecx
@@ -60,7 +60,7 @@ L_more_than_31:
 	movl	$0,8(%esi)
 	popl	%esi
 	leave
-	ret
+	pax_ret shr_Xsig
 
 L_more_than_63:
 	cmpl	$96,%ecx
@@ -75,7 +75,7 @@ L_more_than_63:
 	movl	%edx,8(%esi)
 	popl	%esi
 	leave
-	ret
+	pax_ret shr_Xsig
 
 L_more_than_95:
 	xorl	%eax,%eax
@@ -84,4 +84,4 @@ L_more_than_95:
 	movl	%eax,8(%esi)
 	popl	%esi
 	leave
-	ret
+	pax_ret shr_Xsig
