diff --git a/arch/x86/lib/cmpxchg16b_emu.S b/arch/x86/lib/cmpxchg16b_emu.S
index 9b33024..cdbc725 100644
--- a/arch/x86/lib/cmpxchg16b_emu.S
+++ b/arch/x86/lib/cmpxchg16b_emu.S
@@ -7,6 +7,7 @@
  */
 #include <linux/linkage.h>
 #include <asm/percpu.h>
+#include <asm/alternative-asm.h>
 
 .text
 
@@ -43,11 +44,11 @@ ENTRY(this_cpu_cmpxchg16b_emu)
 
 	popfq
 	mov $1, %al
-	ret
+	pax_ret this_cpu_cmpxchg16b_emu
 
 .Lnot_same:
 	popfq
 	xor %al,%al
-	ret
+	pax_ret this_cpu_cmpxchg16b_emu
 
 ENDPROC(this_cpu_cmpxchg16b_emu)
