diff --git a/arch/mips/lib/libgcc.h b/arch/mips/lib/libgcc.h
index 05909d58..b03284b 100644
--- a/arch/mips/lib/libgcc.h
+++ b/arch/mips/lib/libgcc.h
@@ -5,9 +5,15 @@
 
 typedef int word_type __attribute__ ((mode (__word__)));
 
+#ifdef CONFIG_64BIT
+typedef int DWtype __attribute__((mode(TI)));
+#else
+typedef long long DWtype;
+#endif
+
 #ifdef __BIG_ENDIAN
 struct DWstruct {
-	int high, low;
+	long high, low;
 };
 
 struct TWstruct {
@@ -15,7 +21,7 @@ struct TWstruct {
 };
 #elif defined(__LITTLE_ENDIAN)
 struct DWstruct {
-	int low, high;
+	long low, high;
 };
 
 struct TWstruct {
@@ -27,7 +33,7 @@ struct TWstruct {
 
 typedef union {
 	struct DWstruct s;
-	long long ll;
+	DWtype ll;
 } DWunion;
 
 #if defined(CONFIG_64BIT) && defined(CONFIG_CPU_MIPSR6)
