diff --git a/drivers/staging/rtl8192e/rtl8192e/rtl_core.h b/drivers/staging/rtl8192e/rtl8192e/rtl_core.h
index babc0b3..2680a8b 100644
--- a/drivers/staging/rtl8192e/rtl8192e/rtl_core.h
+++ b/drivers/staging/rtl8192e/rtl8192e/rtl_core.h
@@ -586,12 +586,12 @@ void force_pci_posting(struct net_device *dev);
 void rtl92e_rx_enable(struct net_device *);
 void rtl92e_tx_enable(struct net_device *);
 
-void rtl92e_hw_sleep_wq(void *data);
+void rtl92e_hw_sleep_wq(struct work_struct *data);
 void rtl92e_commit(struct net_device *dev);
 
 void rtl92e_check_rfctrl_gpio_timer(unsigned long data);
 
-void rtl92e_hw_wakeup_wq(void *data);
+void rtl92e_hw_wakeup_wq(struct work_struct *data);
 
 void rtl92e_reset_desc_ring(struct net_device *dev);
 void rtl92e_set_wireless_mode(struct net_device *dev, u8 wireless_mode);
