diff --git a/drivers/net/wireless/realtek/rtlwifi/ps.c b/drivers/net/wireless/realtek/rtlwifi/ps.c
index d0ffc4d..5f197b4 100644
--- a/drivers/net/wireless/realtek/rtlwifi/ps.c
+++ b/drivers/net/wireless/realtek/rtlwifi/ps.c
@@ -198,7 +198,7 @@ static void _rtl_ps_inactive_ps(struct ieee80211_hw *hw)
 	ppsc->swrf_processing = false;
 }
 
-void rtl_ips_nic_off_wq_callback(void *data)
+void rtl_ips_nic_off_wq_callback(struct work_struct *data)
 {
 	struct rtl_works *rtlworks =
 	    container_of_dwork_rtl(data, struct rtl_works, ips_nic_off_wq);
@@ -582,7 +582,7 @@ void rtl_swlps_rf_awake(struct ieee80211_hw *hw)
 	spin_unlock_irqrestore(&rtlpriv->locks.lps_lock, flag);
 }
 
-void rtl_swlps_rfon_wq_callback(void *data)
+void rtl_swlps_rfon_wq_callback(struct work_struct *data)
 {
 	struct rtl_works *rtlworks =
 	    container_of_dwork_rtl(data, struct rtl_works, ps_rfon_wq);
@@ -696,7 +696,7 @@ void rtl_lps_leave(struct ieee80211_hw *hw)
 }
 EXPORT_SYMBOL_GPL(rtl_lps_leave);
 
-void rtl_swlps_wq_callback(void *data)
+void rtl_swlps_wq_callback(struct work_struct *data)
 {
 	struct rtl_works *rtlworks = container_of_dwork_rtl(data,
 				     struct rtl_works,
