diff --git a/drivers/staging/lustre/lustre/ptlrpc/layout.c b/drivers/staging/lustre/lustre/ptlrpc/layout.c
index 839ef3e..8912e22 100644
--- a/drivers/staging/lustre/lustre/ptlrpc/layout.c
+++ b/drivers/staging/lustre/lustre/ptlrpc/layout.c
@@ -778,8 +778,8 @@ struct req_capsule;
 	.rmf_name    = (name),					\
 	.rmf_flags   = (flags),					\
 	.rmf_size    = (size),					\
-	.rmf_swabber = (void (*)(void *))(swabber),		\
-	.rmf_dumper  = (void (*)(void *))(dumper)		\
+	.rmf_swabber = (swabber),				\
+	.rmf_dumper  = (dumper)					\
 }
 
 struct req_msg_field RMF_GENERIC_DATA =
@@ -1871,8 +1871,7 @@ static void *__req_capsule_get(struct req_capsule *pill,
 	msg = __req_msg(pill, loc);
 	LASSERT(msg);
 
-	getter = (field->rmf_flags & RMF_F_STRING) ?
-		(typeof(getter))lustre_msg_string : lustre_msg_buf;
+	getter = (field->rmf_flags & RMF_F_STRING) ?  lustre_msg_string : lustre_msg_buf;
 
 	if (field->rmf_flags & RMF_F_STRUCT_ARRAY) {
 		/*
