diff --git a/drivers/net/wireless/ath/ath9k/main.c b/drivers/net/wireless/ath/ath9k/main.c
index b868f02..95db87b 100644
--- a/drivers/net/wireless/ath/ath9k/main.c
+++ b/drivers/net/wireless/ath/ath9k/main.c
@@ -2617,16 +2617,18 @@ void ath9k_fill_chanctx_ops(void)
 	if (!ath9k_is_chanctx_enabled())
 		return;
 
-	ath9k_ops.hw_scan                  = ath9k_hw_scan;
-	ath9k_ops.cancel_hw_scan           = ath9k_cancel_hw_scan;
-	ath9k_ops.remain_on_channel        = ath9k_remain_on_channel;
-	ath9k_ops.cancel_remain_on_channel = ath9k_cancel_remain_on_channel;
-	ath9k_ops.add_chanctx              = ath9k_add_chanctx;
-	ath9k_ops.remove_chanctx           = ath9k_remove_chanctx;
-	ath9k_ops.change_chanctx           = ath9k_change_chanctx;
-	ath9k_ops.assign_vif_chanctx       = ath9k_assign_vif_chanctx;
-	ath9k_ops.unassign_vif_chanctx     = ath9k_unassign_vif_chanctx;
-	ath9k_ops.mgd_prepare_tx           = ath9k_mgd_prepare_tx;
+	pax_open_kernel();
+	const_cast(ath9k_ops.hw_scan)                  = ath9k_hw_scan;
+	const_cast(ath9k_ops.cancel_hw_scan)           = ath9k_cancel_hw_scan;
+	const_cast(ath9k_ops.remain_on_channel)        = ath9k_remain_on_channel;
+	const_cast(ath9k_ops.cancel_remain_on_channel) = ath9k_cancel_remain_on_channel;
+	const_cast(ath9k_ops.add_chanctx)              = ath9k_add_chanctx;
+	const_cast(ath9k_ops.remove_chanctx)           = ath9k_remove_chanctx;
+	const_cast(ath9k_ops.change_chanctx)           = ath9k_change_chanctx;
+	const_cast(ath9k_ops.assign_vif_chanctx)       = ath9k_assign_vif_chanctx;
+	const_cast(ath9k_ops.unassign_vif_chanctx)     = ath9k_unassign_vif_chanctx;
+	const_cast(ath9k_ops.mgd_prepare_tx)           = ath9k_mgd_prepare_tx;
+	pax_close_kernel();
 }
 
 #endif
