diff --git a/drivers/net/ethernet/nvidia/forcedeth.c b/drivers/net/ethernet/nvidia/forcedeth.c
index 9b0d7f4..c29155f 100644
--- a/drivers/net/ethernet/nvidia/forcedeth.c
+++ b/drivers/net/ethernet/nvidia/forcedeth.c
@@ -357,8 +357,8 @@ struct ring_desc {
 };
 
 struct ring_desc_ex {
-	__le32 bufhigh;
-	__le32 buflow;
+	__le32 bufhigh __intentional_overflow(0);
+	__le32 buflow __intentional_overflow(0);
 	__le32 txvlan;
 	__le32 flaglen;
 };
