diff --git a/drivers/net/can/led.c b/drivers/net/can/led.c
index c1b6676..50a8a51 100644
--- a/drivers/net/can/led.c
+++ b/drivers/net/can/led.c
@@ -128,7 +128,7 @@ static int can_led_notifier(struct notifier_block *nb, unsigned long msg,
 }
 
 /* notifier block for netdevice event */
-static struct notifier_block can_netdev_notifier __read_mostly = {
+static struct notifier_block can_netdev_notifier = {
 	.notifier_call = can_led_notifier,
 };
 
