diff --git a/drivers/media/platform/soc_camera/soc_camera.c b/drivers/media/platform/soc_camera/soc_camera.c
index edd1c1d..fbec29e 100644
--- a/drivers/media/platform/soc_camera/soc_camera.c
+++ b/drivers/media/platform/soc_camera/soc_camera.c
@@ -1723,7 +1723,7 @@ static int soc_camera_probe(struct soc_camera_host *ici,
 			goto eadd;
 
 		if (shd->module_name)
-			ret = request_module(shd->module_name);
+			ret = request_module("%s", shd->module_name);
 
 		ret = shd->add_device(icd);
 		if (ret < 0)
