diff --git a/drivers/net/ethernet/apm/xgene/xgene_enet_main.c b/drivers/net/ethernet/apm/xgene/xgene_enet_main.c
index 8158d46..f22bcfc 100644
--- a/drivers/net/ethernet/apm/xgene/xgene_enet_main.c
+++ b/drivers/net/ethernet/apm/xgene/xgene_enet_main.c
@@ -122,7 +122,7 @@ static void xgene_enet_delete_bufpool(struct xgene_enet_desc_ring *buf_pool)
 	}
 }
 
-static irqreturn_t xgene_enet_rx_irq(const int irq, void *data)
+static irqreturn_t xgene_enet_rx_irq(int irq, void *data)
 {
 	struct xgene_enet_desc_ring *rx_ring = data;
 
@@ -642,7 +642,7 @@ static int xgene_enet_process_ring(struct xgene_enet_desc_ring *ring,
 	return processed;
 }
 
-static int xgene_enet_napi(struct napi_struct *napi, const int budget)
+static int xgene_enet_napi(struct napi_struct *napi, int budget)
 {
 	struct xgene_enet_desc_ring *ring;
 	int processed;
