diff --git a/drivers/staging/unisys/visorbus/visorchipset.c b/drivers/staging/unisys/visorbus/visorchipset.c
index 5987149..15c44ea 100644
--- a/drivers/staging/unisys/visorbus/visorchipset.c
+++ b/drivers/staging/unisys/visorbus/visorchipset.c
@@ -2234,7 +2234,7 @@ static __init uint32_t visorutil_spar_detect(void)
 	}
 }
 
-static int init_unisys(void)
+static __init int init_unisys(void)
 {
 	int result;
 
@@ -2249,7 +2249,7 @@ static int init_unisys(void)
 	return 0;
 };
 
-static void exit_unisys(void)
+static __exit void exit_unisys(void)
 {
 	acpi_bus_unregister_driver(&unisys_acpi_driver);
 }
