diff --git a/drivers/gpu/drm/amd/scheduler/gpu_scheduler.c b/drivers/gpu/drm/amd/scheduler/gpu_scheduler.c
index ffe1f85..7017bfc 100644
--- a/drivers/gpu/drm/amd/scheduler/gpu_scheduler.c
+++ b/drivers/gpu/drm/amd/scheduler/gpu_scheduler.c
@@ -137,7 +137,7 @@ int amd_sched_entity_init(struct amd_gpu_scheduler *sched,
 	if (r)
 		return r;
 
-	atomic_set(&entity->fence_seq, 0);
+	atomic_set_unchecked(&entity->fence_seq, 0);
 	entity->fence_context = fence_context_alloc(2);
 
 	return 0;
