diff --git a/drivers/infiniband/core/netlink.c b/drivers/infiniband/core/netlink.c
index 10469b0..e8b45f3 100644
--- a/drivers/infiniband/core/netlink.c
+++ b/drivers/infiniband/core/netlink.c
@@ -176,11 +176,10 @@ static int ibnl_rcv_msg(struct sk_buff *skb, struct nlmsghdr *nlh)
 			}
 
 			{
-				struct netlink_dump_control c = {
+				netlink_dump_control_no_const c = {
 					.dump = client->cb_table[op].dump,
-					.module = client->cb_table[op].module,
 				};
-				return netlink_dump_start(nls, skb, nlh, &c);
+				return __netlink_dump_start(nls, skb, nlh, &c, NULL, client->cb_table[op].module);
 			}
 		}
 	}
