diff --git a/drivers/pci/hotplug/cpqphp_nvram.c b/drivers/pci/hotplug/cpqphp_nvram.c
index c25fc90..b054774 100644
--- a/drivers/pci/hotplug/cpqphp_nvram.c
+++ b/drivers/pci/hotplug/cpqphp_nvram.c
@@ -425,8 +425,10 @@ static u32 store_HRT(void __iomem *rom_start)
 
 void compaq_nvram_init(void __iomem *rom_start)
 {
+#ifndef CONFIG_PAX_KERNEXEC
 	if (rom_start)
 		compaq_int15_entry_point = (rom_start + ROM_INT15_PHY_ADDR - ROM_PHY_ADDR);
+#endif
 
 	dbg("int15 entry  = %p\n", compaq_int15_entry_point);
 
