diff --git a/drivers/pci/msi.c b/drivers/pci/msi.c
index 3455f75..889b244 100644
--- a/drivers/pci/msi.c
+++ b/drivers/pci/msi.c
@@ -477,8 +477,8 @@ static int populate_msi_sysfs(struct pci_dev *pdev)
 {
 	struct attribute **msi_attrs;
 	struct attribute *msi_attr;
-	struct device_attribute *msi_dev_attr;
-	struct attribute_group *msi_irq_group;
+	device_attribute_no_const *msi_dev_attr;
+	attribute_group_no_const *msi_irq_group;
 	const struct attribute_group **msi_irq_groups;
 	struct msi_desc *entry;
 	int ret = -ENOMEM;
@@ -540,7 +540,7 @@ static int populate_msi_sysfs(struct pci_dev *pdev)
 	count = 0;
 	msi_attr = msi_attrs[count];
 	while (msi_attr) {
-		msi_dev_attr = container_of(msi_attr, struct device_attribute, attr);
+		msi_dev_attr = container_of(msi_attr, device_attribute_no_const, attr);
 		kfree(msi_attr->name);
 		kfree(msi_dev_attr);
 		++count;
@@ -1418,12 +1418,14 @@ static void pci_msi_domain_update_dom_ops(struct msi_domain_info *info)
 	if (ops == NULL) {
 		info->ops = &pci_msi_domain_ops_default;
 	} else {
+		pax_open_kernel();
 		if (ops->set_desc == NULL)
-			ops->set_desc = pci_msi_domain_set_desc;
+			const_cast(ops->set_desc) = pci_msi_domain_set_desc;
 		if (ops->msi_check == NULL)
-			ops->msi_check = pci_msi_domain_check_cap;
+			const_cast(ops->msi_check) = pci_msi_domain_check_cap;
 		if (ops->handle_error == NULL)
-			ops->handle_error = pci_msi_domain_handle_error;
+			const_cast(ops->handle_error) = pci_msi_domain_handle_error;
+		pax_close_kernel();
 	}
 }
 
@@ -1432,12 +1434,14 @@ static void pci_msi_domain_update_chip_ops(struct msi_domain_info *info)
 	struct irq_chip *chip = info->chip;
 
 	BUG_ON(!chip);
+	pax_open_kernel();
 	if (!chip->irq_write_msi_msg)
-		chip->irq_write_msi_msg = pci_msi_domain_write_msg;
+		const_cast(chip->irq_write_msi_msg) = pci_msi_domain_write_msg;
 	if (!chip->irq_mask)
-		chip->irq_mask = pci_msi_mask_irq;
+		const_cast(chip->irq_mask) = pci_msi_mask_irq;
 	if (!chip->irq_unmask)
-		chip->irq_unmask = pci_msi_unmask_irq;
+		const_cast(chip->irq_unmask) = pci_msi_unmask_irq;
+	pax_close_kernel();
 }
 
 /**
