diff --git a/drivers/net/vrf.c b/drivers/net/vrf.c
index a2afb8e..6d66a2e 100644
--- a/drivers/net/vrf.c
+++ b/drivers/net/vrf.c
@@ -1299,7 +1299,7 @@ static const struct nla_policy vrf_nl_policy[IFLA_VRF_MAX + 1] = {
 	[IFLA_VRF_TABLE] = { .type = NLA_U32 },
 };
 
-static struct rtnl_link_ops vrf_link_ops __read_mostly = {
+static struct rtnl_link_ops vrf_link_ops = {
 	.kind		= DRV_NAME,
 	.priv_size	= sizeof(struct net_vrf),
 
@@ -1336,7 +1336,7 @@ static int vrf_device_event(struct notifier_block *unused,
 	return NOTIFY_DONE;
 }
 
-static struct notifier_block vrf_notifier_block __read_mostly = {
+static struct notifier_block vrf_notifier_block = {
 	.notifier_call = vrf_device_event,
 };
 
