diff --git a/drivers/staging/lustre/lustre/llite/dir.c b/drivers/staging/lustre/lustre/llite/dir.c
index 7f32a53..234abb7 100644
--- a/drivers/staging/lustre/lustre/llite/dir.c
+++ b/drivers/staging/lustre/lustre/llite/dir.c
@@ -136,11 +136,12 @@
 struct page *ll_get_dir_page(struct inode *dir, struct md_op_data *op_data,
 			     __u64 offset)
 {
-	struct md_callback cb_op;
+	static struct md_callback cb_op = {
+		.md_blocking_ast = ll_md_blocking_ast,
+	};
 	struct page *page;
 	int rc;
 
-	cb_op.md_blocking_ast = ll_md_blocking_ast;
 	rc = md_read_page(ll_i2mdexp(dir), op_data, &cb_op, offset, &page);
 	if (rc)
 		return ERR_PTR(rc);
