diff --git a/drivers/firmware/psci.c b/drivers/firmware/psci.c
index 8263429..d0ef61f 100644
--- a/drivers/firmware/psci.c
+++ b/drivers/firmware/psci.c
@@ -62,7 +62,7 @@ bool psci_tos_resident_on(int cpu)
 struct psci_operations psci_ops = {
 	.conduit = PSCI_CONDUIT_NONE,
 	.smccc_version = SMCCC_VERSION_1_0,
-};
+} __read_only;
 
 typedef unsigned long (psci_fn)(unsigned long, unsigned long,
 				unsigned long, unsigned long);
