diff --git a/drivers/mfd/rn5t618.c b/drivers/mfd/rn5t618.c
index ee94080..e2a4a3d 100644
--- a/drivers/mfd/rn5t618.c
+++ b/drivers/mfd/rn5t618.c
@@ -52,7 +52,6 @@ static const struct regmap_config rn5t618_regmap_config = {
 };
 
 static struct rn5t618 *rn5t618_pm_power_off;
-static struct notifier_block rn5t618_restart_handler;
 
 static void rn5t618_trigger_poweroff_sequence(bool repower)
 {
@@ -84,6 +83,12 @@ static int rn5t618_restart(struct notifier_block *this,
 	return NOTIFY_DONE;
 }
 
+static struct notifier_block rn5t618_restart_handler = {
+	.notifier_call = rn5t618_restart,
+	.priority = 192,
+
+};
+
 static const struct of_device_id rn5t618_of_match[] = {
 	{ .compatible = "ricoh,rn5t567", .data = (void *)RN5T567 },
 	{ .compatible = "ricoh,rn5t618", .data = (void *)RN5T618 },
@@ -133,9 +138,6 @@ static int rn5t618_i2c_probe(struct i2c_client *i2c,
 			dev_warn(&i2c->dev, "Poweroff callback already assigned\n");
 	}
 
-	rn5t618_restart_handler.notifier_call = rn5t618_restart;
-	rn5t618_restart_handler.priority = 192;
-
 	ret = register_restart_handler(&rn5t618_restart_handler);
 	if (ret) {
 		dev_err(&i2c->dev, "cannot register restart handler, %d\n", ret);
