diff --git a/drivers/gpio/gpiolib.c b/drivers/gpio/gpiolib.c
index 9215931..8e92a9e 100644
--- a/drivers/gpio/gpiolib.c
+++ b/drivers/gpio/gpiolib.c
@@ -1675,8 +1675,10 @@ static void gpiochip_irqchip_remove(struct gpio_chip *gpiochip)
 	}
 
 	if (gpiochip->irqchip) {
-		gpiochip->irqchip->irq_request_resources = NULL;
-		gpiochip->irqchip->irq_release_resources = NULL;
+		pax_open_kernel();
+		const_cast(gpiochip->irqchip->irq_request_resources) = NULL;
+		const_cast(gpiochip->irqchip->irq_release_resources) = NULL;
+		pax_close_kernel();
 		gpiochip->irqchip = NULL;
 	}
 
@@ -1770,8 +1772,10 @@ int _gpiochip_irqchip_add(struct gpio_chip *gpiochip,
 	 */
 	if (!irqchip->irq_request_resources &&
 	    !irqchip->irq_release_resources) {
-		irqchip->irq_request_resources = gpiochip_irq_reqres;
-		irqchip->irq_release_resources = gpiochip_irq_relres;
+		pax_open_kernel();
+		const_cast(irqchip->irq_request_resources) = gpiochip_irq_reqres;
+		const_cast(irqchip->irq_release_resources) = gpiochip_irq_relres;
+		pax_close_kernel();
 	}
 
 	/*
