diff --git a/drivers/android/binder.c b/drivers/android/binder.c
index 3c71b98..03bc39b9 100644
--- a/drivers/android/binder.c
+++ b/drivers/android/binder.c
@@ -119,7 +119,7 @@ static DECLARE_WAIT_QUEUE_HEAD(binder_user_error_wait);
 static int binder_stop_on_user_error;
 
 static int binder_set_stop_on_user_error(const char *val,
-					 struct kernel_param *kp)
+					 const struct kernel_param *kp)
 {
 	int ret;
 
