diff --git a/drivers/regulator/max8660.c b/drivers/regulator/max8660.c
index b87f62d..34f1cdf 100644
--- a/drivers/regulator/max8660.c
+++ b/drivers/regulator/max8660.c
@@ -423,8 +423,10 @@ static int max8660_probe(struct i2c_client *client,
 		max8660->shadow_regs[MAX8660_OVER1] = 5;
 	} else {
 		/* Otherwise devices can be toggled via software */
-		max8660_dcdc_ops.enable = max8660_dcdc_enable;
-		max8660_dcdc_ops.disable = max8660_dcdc_disable;
+		pax_open_kernel();
+		const_cast(max8660_dcdc_ops.enable) = max8660_dcdc_enable;
+		const_cast(max8660_dcdc_ops.disable) = max8660_dcdc_disable;
+		pax_close_kernel();
 	}
 
 	/*
