diff --git a/drivers/base/power/domain.c b/drivers/base/power/domain.c
index e023066..a28458a 100644
--- a/drivers/base/power/domain.c
+++ b/drivers/base/power/domain.c
@@ -1895,8 +1895,10 @@ int genpd_dev_pm_attach(struct device *dev)
 		goto out;
 	}
 
-	dev->pm_domain->detach = genpd_dev_pm_detach;
-	dev->pm_domain->sync = genpd_dev_pm_sync;
+	pax_open_kernel();
+	const_cast(dev->pm_domain->detach) = genpd_dev_pm_detach;
+	const_cast(dev->pm_domain->sync) = genpd_dev_pm_sync;
+	pax_close_kernel();
 
 	mutex_lock(&pd->lock);
 	ret = genpd_poweron(pd, 0);
