diff --git a/drivers/md/bcache/writeback.c b/drivers/md/bcache/writeback.c
index e51644e..5cc3c15 100644
--- a/drivers/md/bcache/writeback.c
+++ b/drivers/md/bcache/writeback.c
@@ -117,14 +117,16 @@ static void dirty_init(struct keybuf_key *w)
 	bch_bio_map(bio, NULL);
 }
 
-static void dirty_io_destructor(struct closure *cl)
+static void dirty_io_destructor(struct work_struct *work)
 {
+	struct closure *cl = container_of(work, struct closure, work);
 	struct dirty_io *io = container_of(cl, struct dirty_io, cl);
 	kfree(io);
 }
 
-static void write_dirty_finish(struct closure *cl)
+static void write_dirty_finish(struct work_struct *work)
 {
+	struct closure *cl = container_of(work, struct closure, work);
 	struct dirty_io *io = container_of(cl, struct dirty_io, cl);
 	struct keybuf_key *w = io->bio.bi_private;
 	struct cached_dev *dc = io->dc;
@@ -173,8 +175,9 @@ static void dirty_endio(struct bio *bio)
 	closure_put(&io->cl);
 }
 
-static void write_dirty(struct closure *cl)
+static void write_dirty(struct work_struct *work)
 {
+	struct closure *cl = container_of(work, struct closure, work);
 	struct dirty_io *io = container_of(cl, struct dirty_io, cl);
 	struct keybuf_key *w = io->bio.bi_private;
 
@@ -200,8 +203,9 @@ static void read_dirty_endio(struct bio *bio)
 	dirty_endio(bio);
 }
 
-static void read_dirty_submit(struct closure *cl)
+static void read_dirty_submit(struct work_struct *work)
 {
+	struct closure *cl = container_of(work, struct closure, work);
 	struct dirty_io *io = container_of(cl, struct dirty_io, cl);
 
 	closure_bio_submit(&io->bio, cl);
