diff --git a/drivers/char/Kconfig b/drivers/char/Kconfig
index 8453a49..aaea2ac 100644
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -17,7 +17,8 @@ config DEVMEM
 
 config DEVKMEM
 	bool "/dev/kmem virtual device support"
-	default y
+	default n
+	depends on !GRKERNSEC_KMEM
 	help
 	  Say Y here if you want to support the /dev/kmem device. The
 	  /dev/kmem device is rarely used, but can be used for certain
@@ -573,6 +574,7 @@ config TELCLOCK
 config DEVPORT
 	bool "/dev/port character device"
 	depends on ISA || PCI
+	depends on !GRKERNSEC_KMEM
 	default y
 	help
 	  Say Y here if you want to support the /dev/port device. The /dev/port
