diff --git a/drivers/base/regmap/regmap.c b/drivers/base/regmap/regmap.c
index ae63bb0..6481579 100644
--- a/drivers/base/regmap/regmap.c
+++ b/drivers/base/regmap/regmap.c
@@ -425,8 +425,8 @@ static void regmap_unlock_mutex(void *__map)
 	mutex_unlock(&map->mutex);
 }
 
+static void regmap_lock_spinlock(void *__map) __acquires(&map->spinlock);
 static void regmap_lock_spinlock(void *__map)
-__acquires(&map->spinlock)
 {
 	struct regmap *map = __map;
 	unsigned long flags;
@@ -435,8 +435,8 @@ __acquires(&map->spinlock)
 	map->spinlock_flags = flags;
 }
 
+static void regmap_unlock_spinlock(void *__map) __releases(&map->spinlock);
 static void regmap_unlock_spinlock(void *__map)
-__releases(&map->spinlock)
 {
 	struct regmap *map = __map;
 	spin_unlock_irqrestore(&map->spinlock, map->spinlock_flags);
