diff --git a/drivers/gpu/drm/sti/sti_hqvdp.c b/drivers/gpu/drm/sti/sti_hqvdp.c
index f88130f..a2be8a4 100644
--- a/drivers/gpu/drm/sti/sti_hqvdp.c
+++ b/drivers/gpu/drm/sti/sti_hqvdp.c
@@ -628,7 +628,7 @@ static int hqvdp_dbg_show(struct seq_file *s, void *data)
 	return 0;
 }
 
-static struct drm_info_list hqvdp_debugfs_files[] = {
+static drm_info_list_no_const hqvdp_debugfs_files[] __read_only = {
 	{ "hqvdp", hqvdp_dbg_show, 0, NULL },
 };
 
@@ -636,8 +636,10 @@ static int hqvdp_debugfs_init(struct sti_hqvdp *hqvdp, struct drm_minor *minor)
 {
 	unsigned int i;
 
+	pax_open_kernel();
 	for (i = 0; i < ARRAY_SIZE(hqvdp_debugfs_files); i++)
 		hqvdp_debugfs_files[i].data = hqvdp;
+	pax_close_kernel();
 
 	return drm_debugfs_create_files(hqvdp_debugfs_files,
 					ARRAY_SIZE(hqvdp_debugfs_files),
