diff --git a/drivers/clk/ti/clk.c b/drivers/clk/ti/clk.c
index 5fcf247..446780a 100644
--- a/drivers/clk/ti/clk.c
+++ b/drivers/clk/ti/clk.c
@@ -25,6 +25,8 @@
 #include <linux/regmap.h>
 #include <linux/bootmem.h>
 
+#include <asm/pgtable.h>
+
 #include "clock.h"
 
 #undef pr_fmt
@@ -84,8 +86,10 @@ int ti_clk_setup_ll_ops(struct ti_clk_ll_ops *ops)
 	}
 
 	ti_clk_ll_ops = ops;
-	ops->clk_readl = clk_memmap_readl;
-	ops->clk_writel = clk_memmap_writel;
+	pax_open_kernel();
+	const_cast(ops->clk_readl) = clk_memmap_readl;
+	const_cast(ops->clk_writel) = clk_memmap_writel;
+	pax_close_kernel();
 
 	return 0;
 }
