diff --git a/drivers/staging/lustre/lustre/llite/statahead.c b/drivers/staging/lustre/lustre/llite/statahead.c
index 0677513..4c7ea03 100644
--- a/drivers/staging/lustre/lustre/llite/statahead.c
+++ b/drivers/staging/lustre/lustre/llite/statahead.c
@@ -893,7 +893,7 @@ static int ll_agl_thread(void *arg)
 	CDEBUG(D_READA, "agl thread started: sai %p, parent %pd\n",
 	       sai, parent);
 
-	atomic_inc(&sbi->ll_agl_total);
+	atomic_inc_unchecked(&sbi->ll_agl_total);
 	spin_lock(&plli->lli_agl_lock);
 	sai->sai_agl_valid = 1;
 	if (thread_is_init(thread))
@@ -1010,7 +1010,7 @@ static int ll_statahead_thread(void *arg)
 	if (sbi->ll_flags & LL_SBI_AGL_ENABLED)
 		ll_start_agl(parent, sai);
 
-	atomic_inc(&sbi->ll_sa_total);
+	atomic_inc_unchecked(&sbi->ll_sa_total);
 	spin_lock(&lli->lli_sa_lock);
 	if (thread_is_init(sa_thread))
 		/* If someone else has changed the thread state
@@ -1123,7 +1123,7 @@ static int ll_statahead_thread(void *arg)
 
 		if (sa_low_hit(sai)) {
 			rc = -EFAULT;
-			atomic_inc(&sbi->ll_sa_wrong);
+			atomic_inc_unchecked(&sbi->ll_sa_wrong);
 			CDEBUG(D_READA, "Statahead for dir "DFID" hit ratio too low: hit/miss %llu/%llu, sent/replied %llu/%llu, stopping statahead thread: pid %d\n",
 			       PFID(&lli->lli_fid), sai->sai_hit,
 			       sai->sai_miss, sai->sai_sent,
