diff --git a/fs/nls/nls_koi8-ru.c b/fs/nls/nls_koi8-ru.c
index a80a741..f28c9c9 100644
--- a/fs/nls/nls_koi8-ru.c
+++ b/fs/nls/nls_koi8-ru.c
@@ -13,7 +13,7 @@
 
 static struct nls_table *p_nls;
 
-static int uni2char(const wchar_t uni,
+static int uni2char(wchar_t uni,
 		    unsigned char *out, int boundlen)
 {
 	if (boundlen <= 0)
@@ -62,8 +62,10 @@ static int __init init_nls_koi8_ru(void)
 	p_nls = load_nls("koi8-u");
 
 	if (p_nls) {
-		table.charset2upper = p_nls->charset2upper;
-		table.charset2lower = p_nls->charset2lower;
+		pax_open_kernel();
+		const_cast(table.charset2upper) = p_nls->charset2upper;
+		const_cast(table.charset2lower) = p_nls->charset2lower;
+		pax_close_kernel();
 		return register_nls(&table);
 	}
 
