diff --git a/fs/super.c b/fs/super.c
index 1058bf3..c3e0b54 100644
--- a/fs/super.c
+++ b/fs/super.c
@@ -357,7 +357,8 @@ EXPORT_SYMBOL(deactivate_super);
  *	called for superblocks not in rundown mode (== ones still on ->fs_supers
  *	of their type), so increment of ->s_count is OK here.
  */
-static int grab_super(struct super_block *s) __releases(sb_lock)
+static int grab_super(struct super_block *s) __releases(&sb_lock);
+static int grab_super(struct super_block *s)
 {
 	s->s_count++;
 	spin_unlock(&sb_lock);
