diff --git a/fs/userfaultfd.c b/fs/userfaultfd.c
index 85959d8..6e511a7 100644
--- a/fs/userfaultfd.c
+++ b/fs/userfaultfd.c
@@ -432,7 +432,7 @@ static int userfaultfd_release(struct inode *inode, struct file *file)
 	struct userfaultfd_wake_range range = { .len = 0, };
 	unsigned long new_flags;
 
-	ACCESS_ONCE(ctx->released) = true;
+	ACCESS_ONCE_RW(ctx->released) = true;
 
 	if (!mmget_not_zero(mm))
 		goto wakeup;
