diff --git a/fs/ext4/xattr.c b/fs/ext4/xattr.c
index 3eeed8f..d68ad95 100644
--- a/fs/ext4/xattr.c
+++ b/fs/ext4/xattr.c
@@ -409,7 +409,7 @@ static int
 ext4_xattr_list_entries(struct dentry *dentry, struct ext4_xattr_entry *entry,
 			char *buffer, size_t buffer_size)
 {
-	size_t rest = buffer_size;
+	size_t rest = buffer_size, total_size = 0;
 
 	for (; !IS_LAST_ENTRY(entry); entry = EXT4_XATTR_NEXT(entry)) {
 		const struct xattr_handler *handler =
@@ -430,9 +430,10 @@ ext4_xattr_list_entries(struct dentry *dentry, struct ext4_xattr_entry *entry,
 				*buffer++ = 0;
 			}
 			rest -= size;
+			total_size += size;
 		}
 	}
-	return buffer_size - rest;  /* total size */
+	return total_size;
 }
 
 static int
