diff --git a/fs/block_dev.c b/fs/block_dev.c
index 092a2ee..8f6433d 100644
--- a/fs/block_dev.c
+++ b/fs/block_dev.c
@@ -835,7 +835,7 @@ static bool bd_may_claim(struct block_device *bdev, struct block_device *whole,
 	else if (whole == bdev)
 		return true;  	 /* is a whole device which isn't held */
 
-	else if (whole->bd_holder == bd_may_claim)
+	else if (whole->bd_holder == (void *)bd_may_claim)
 		return true; 	 /* is a partition of a device that is being partitioned */
 	else if (whole->bd_holder != NULL)
 		return false;	 /* is a partition of a held device */
