diff --git a/fs/f2fs/super.c b/fs/f2fs/super.c
index 013c6a5..6913b82 100644
--- a/fs/f2fs/super.c
+++ b/fs/f2fs/super.c
@@ -59,7 +59,7 @@ static void f2fs_build_fault_attr(struct f2fs_sb_info *sbi,
 	struct f2fs_fault_info *ffi = &sbi->fault_info;
 
 	if (rate) {
-		atomic_set(&ffi->inject_ops, 0);
+		atomic_set_unchecked(&ffi->inject_ops, 0);
 		ffi->inject_rate = rate;
 		ffi->inject_type = (1 << FAULT_MAX) - 1;
 	} else {
