diff --git a/fs/btrfs/tree-log.c b/fs/btrfs/tree-log.c
index b890045..ebd25f8 100644
--- a/fs/btrfs/tree-log.c
+++ b/fs/btrfs/tree-log.c
@@ -174,7 +174,7 @@ static int start_log_trans(struct btrfs_trans_handle *trans,
 		root->log_start_pid = current->pid;
 	}
 
-	atomic_inc(&root->log_batch);
+	atomic_inc_unchecked(&root->log_batch);
 	atomic_inc(&root->log_writers);
 	if (ctx) {
 		int index = root->log_transid % 2;
@@ -2768,7 +2768,7 @@ int btrfs_sync_log(struct btrfs_trans_handle *trans,
 		wait_log_commit(root, log_transid - 1);
 
 	while (1) {
-		int batch = atomic_read(&root->log_batch);
+		int batch = atomic_read_unchecked(&root->log_batch);
 		/* when we're on an ssd, just kick the log commit out */
 		if (!btrfs_test_opt(root->fs_info, SSD) &&
 		    test_bit(BTRFS_ROOT_MULTI_LOG_TASKS, &root->state)) {
@@ -2777,7 +2777,7 @@ int btrfs_sync_log(struct btrfs_trans_handle *trans,
 			mutex_lock(&root->log_mutex);
 		}
 		wait_for_writer(root);
-		if (batch == atomic_read(&root->log_batch))
+		if (batch == atomic_read_unchecked(&root->log_batch))
 			break;
 	}
 
@@ -2823,7 +2823,7 @@ int btrfs_sync_log(struct btrfs_trans_handle *trans,
 	btrfs_init_log_ctx(&root_log_ctx, NULL);
 
 	mutex_lock(&log_root_tree->log_mutex);
-	atomic_inc(&log_root_tree->log_batch);
+	atomic_inc_unchecked(&log_root_tree->log_batch);
 	atomic_inc(&log_root_tree->log_writers);
 
 	index2 = log_root_tree->log_transid % 2;
