diff --git a/fs/jfs/super.c b/fs/jfs/super.c
index 85671f7..995bb95 100644
--- a/fs/jfs/super.c
+++ b/fs/jfs/super.c
@@ -897,8 +897,10 @@ static int __init init_jfs_fs(void)
 	int rc;
 
 	jfs_inode_cachep =
-	    kmem_cache_create("jfs_ip", sizeof(struct jfs_inode_info), 0,
+	    kmem_cache_create_usercopy("jfs_ip", sizeof(struct jfs_inode_info), 0,
 			    SLAB_RECLAIM_ACCOUNT|SLAB_MEM_SPREAD|SLAB_ACCOUNT,
+			    offsetof(struct jfs_inode_info, i_inline),
+			    sizeof(((struct jfs_inode_info *)0)->i_inline),
 			    init_once);
 	if (jfs_inode_cachep == NULL)
 		return -ENOMEM;
