diff --git a/fs/fcntl.c b/fs/fcntl.c
index 350a2c8..9fb9bf7 100644
--- a/fs/fcntl.c
+++ b/fs/fcntl.c
@@ -103,6 +103,10 @@ void __f_setown(struct file *filp, struct pid *pid, enum pid_type type,
 		int force)
 {
 	security_file_set_fowner(filp);
+	if (gr_handle_chroot_fowner(pid, type))
+		return;
+	if (gr_check_protected_task_fowner(pid, type))
+		return;
 	f_modown(filp, pid, type, force);
 }
 EXPORT_SYMBOL(__f_setown);
