diff --git a/fs/nls/nls_cp949.c b/fs/nls/nls_cp949.c
index 199171e..709af9a 100644
--- a/fs/nls/nls_cp949.c
+++ b/fs/nls/nls_cp949.c
@@ -13858,7 +13858,7 @@ static const unsigned char charset2upper[256] = {
 	0xf8, 0xf9, 0xfa, 0xfb, 0xfc, 0xfd, 0xfe, 0xff, /* 0xf8-0xff */
 };
 
-static int uni2char(const wchar_t uni,
+static int uni2char(wchar_t uni,
 			unsigned char *out, int boundlen)
 {
 	const unsigned char *uni2charset;
