diff --git a/fs/coda/dir.c b/fs/coda/dir.c
index c0474ac..2c8be82 100644
--- a/fs/coda/dir.c
+++ b/fs/coda/dir.c
@@ -29,11 +29,10 @@
 #include "coda_int.h"
 
 /* same as fs/bad_inode.c */
-static int coda_return_EIO(void)
+static int coda_mknod(struct inode *inode, struct dentry *dentry, umode_t mode, dev_t dev)
 {
 	return -EIO;
 }
-#define CODA_EIO_ERROR ((void *) (coda_return_EIO))
 
 /* inode operations for directories */
 /* access routines: lookup, readlink, permission */
@@ -572,7 +571,7 @@ const struct inode_operations coda_dir_inode_operations = {
 	.symlink	= coda_symlink,
 	.mkdir		= coda_mkdir,
 	.rmdir		= coda_rmdir,
-	.mknod		= CODA_EIO_ERROR,
+	.mknod		= coda_mknod,
 	.rename		= coda_rename,
 	.permission	= coda_permission,
 	.getattr	= coda_getattr,
