diff --git a/fs/ntfs/inode.h b/fs/ntfs/inode.h
index b3c3469..17208ad 100644
--- a/fs/ntfs/inode.h
+++ b/fs/ntfs/inode.h
@@ -267,9 +267,7 @@ typedef struct {
 	ATTR_TYPE type;
 } ntfs_attr;
 
-typedef int (*test_t)(struct inode *, void *);
-
-extern int ntfs_test_inode(struct inode *vi, ntfs_attr *na);
+extern int ntfs_test_inode(struct inode *vi, void *_na);
 
 extern struct inode *ntfs_iget(struct super_block *sb, unsigned long mft_no);
 extern struct inode *ntfs_attr_iget(struct inode *base_vi, ATTR_TYPE type,
