diff --git a/fs/orangefs/super.c b/fs/orangefs/super.c
index 629d8c9..2edba1f 100644
--- a/fs/orangefs/super.c
+++ b/fs/orangefs/super.c
@@ -587,10 +587,12 @@ void orangefs_kill_sb(struct super_block *sb)
 
 int orangefs_inode_cache_initialize(void)
 {
-	orangefs_inode_cache = kmem_cache_create("orangefs_inode_cache",
+	orangefs_inode_cache = kmem_cache_create_usercopy("orangefs_inode_cache",
 					      sizeof(struct orangefs_inode_s),
 					      0,
 					      ORANGEFS_CACHE_CREATE_FLAGS,
+					      offsetof(struct orangefs_inode_s, link_target),
+					      sizeof(((struct orangefs_inode_s *)0)->link_target),
 					      orangefs_inode_cache_ctor);
 
 	if (!orangefs_inode_cache) {
