diff --git a/fs/btrfs/dev-replace.h b/fs/btrfs/dev-replace.h
index e922b42..2a5a145 100644
--- a/fs/btrfs/dev-replace.h
+++ b/fs/btrfs/dev-replace.h
@@ -42,8 +42,8 @@ void btrfs_dev_replace_set_lock_blocking(struct btrfs_dev_replace *dev_replace);
 void btrfs_dev_replace_clear_lock_blocking(
 					struct btrfs_dev_replace *dev_replace);
 
-static inline void btrfs_dev_replace_stats_inc(atomic64_t *stat_value)
+static inline void btrfs_dev_replace_stats_inc(atomic64_unchecked_t *stat_value)
 {
-	atomic64_inc(stat_value);
+	atomic64_inc_unchecked(stat_value);
 }
 #endif
