diff --git a/fs/btrfs/free-space-cache.h b/fs/btrfs/free-space-cache.h
index 363fdd9..887dbe2 100644
--- a/fs/btrfs/free-space-cache.h
+++ b/fs/btrfs/free-space-cache.h
@@ -48,6 +48,7 @@ struct btrfs_free_space_op {
 	bool (*use_bitmap)(struct btrfs_free_space_ctl *ctl,
 			   struct btrfs_free_space *info);
 };
+typedef struct btrfs_free_space_op __no_const btrfs_free_space_op_no_const;
 
 struct btrfs_io_ctl;
 
