diff --git a/include/sound/compress_driver.h b/include/sound/compress_driver.h
index cee8c00..0ee1834 100644
--- a/include/sound/compress_driver.h
+++ b/include/sound/compress_driver.h
@@ -132,7 +132,7 @@ struct snd_compr_ops {
 			struct snd_compr_caps *caps);
 	int (*get_codec_caps) (struct snd_compr_stream *stream,
 			struct snd_compr_codec_caps *codec);
-};
+} __no_const;
 
 /**
  * struct snd_compr: Compressed device
