diff --git a/include/linux/firewire.h b/include/linux/firewire.h
index d4b7683..9feb066 100644
--- a/include/linux/firewire.h
+++ b/include/linux/firewire.h
@@ -451,7 +451,7 @@ struct fw_iso_context {
 
 struct fw_iso_context *fw_iso_context_create(struct fw_card *card,
 		int type, int channel, int speed, size_t header_size,
-		fw_iso_callback_t callback, void *callback_data);
+		void *callback, void *callback_data);
 int fw_iso_context_set_channels(struct fw_iso_context *ctx, u64 *channels);
 int fw_iso_context_queue(struct fw_iso_context *ctx,
 			 struct fw_iso_packet *packet,
