diff --git a/include/linux/irq.h b/include/linux/irq.h
index 39e3254..650e968 100644
--- a/include/linux/irq.h
+++ b/include/linux/irq.h
@@ -425,7 +425,10 @@ struct irq_chip {
 	void		(*ipi_send_mask)(struct irq_data *data, const struct cpumask *dest);
 
 	unsigned long	flags;
-};
+} __do_const;
+#ifndef _LINUX_IRQDOMAIN_H
+typedef struct irq_chip __no_const irq_chip_no_const;
+#endif
 
 /*
  * irq_chip specific flags
