diff --git a/include/linux/init_task.h b/include/linux/init_task.h
index 325f649..b16f6af 100644
--- a/include/linux/init_task.h
+++ b/include/linux/init_task.h
@@ -244,7 +244,6 @@ extern struct task_group root_task_group;
 	.pending	= {						\
 		.list = LIST_HEAD_INIT(tsk.pending.list),		\
 		.signal = {{0}}},					\
-	.blocked	= {{0}},					\
 	.alloc_lock	= __SPIN_LOCK_UNLOCKED(tsk.alloc_lock),		\
 	.journal_info	= NULL,						\
 	.cpu_timers	= INIT_CPU_TIMERS(tsk.cpu_timers),		\
