diff --git a/include/linux/memory.h b/include/linux/memory.h
index 093607f..9717227 100644
--- a/include/linux/memory.h
+++ b/include/linux/memory.h
@@ -126,7 +126,7 @@ extern struct memory_block *find_memory_block(struct mem_section *);
 
 #ifdef CONFIG_MEMORY_HOTPLUG
 #define hotplug_memory_notifier(fn, pri) ({		\
-	static __meminitdata struct notifier_block fn##_mem_nb =\
+	static __meminitconst struct notifier_block fn##_mem_nb =\
 		{ .notifier_call = fn, .priority = pri };\
 	register_memory_notifier(&fn##_mem_nb);			\
 })
