diff --git a/include/linux/pm.h b/include/linux/pm.h
index 06eb353..dbf4a34 100644
--- a/include/linux/pm.h
+++ b/include/linux/pm.h
@@ -631,6 +631,7 @@ struct dev_pm_domain {
 	void (*sync)(struct device *dev);
 	void (*dismiss)(struct device *dev);
 };
+typedef struct dev_pm_domain __no_const dev_pm_domain_no_const;
 
 /*
  * The PM_EVENT_ messages are also used by drivers implementing the legacy
