diff --git a/include/linux/uaccess.h b/include/linux/uaccess.h
index f30c187..d2b4ce2 100644
--- a/include/linux/uaccess.h
+++ b/include/linux/uaccess.h
@@ -109,7 +109,7 @@ extern long strncpy_from_unsafe(char *dst, const void *unsafe_addr, long count);
  * Returns 0 on success, or -EFAULT.
  */
 #define probe_kernel_address(addr, retval)		\
-	probe_kernel_read(&retval, addr, sizeof(retval))
+	probe_kernel_read(&(retval), addr, sizeof(retval))
 
 #ifndef user_access_begin
 #define user_access_begin() do { } while (0)
