diff --git a/include/linux/fb.h b/include/linux/fb.h
index a964d07..09bf71f 100644
--- a/include/linux/fb.h
+++ b/include/linux/fb.h
@@ -320,7 +320,8 @@ struct fb_ops {
 	/* called at KDB enter and leave time to prepare the console */
 	int (*fb_debug_enter)(struct fb_info *info);
 	int (*fb_debug_leave)(struct fb_info *info);
-};
+} __do_const;
+typedef struct fb_ops __no_const fb_ops_no_const;
 
 #ifdef CONFIG_FB_TILEBLITTING
 #define FB_TILE_CURSOR_NONE        0
