diff --git a/include/linux/if_pppox.h b/include/linux/if_pppox.h
index ba7a9b0..33a0237 100644
--- a/include/linux/if_pppox.h
+++ b/include/linux/if_pppox.h
@@ -78,7 +78,7 @@ struct pppox_proto {
 	int		(*ioctl)(struct socket *sock, unsigned int cmd,
 				 unsigned long arg);
 	struct module	*owner;
-};
+} __do_const;
 
 extern int register_pppox_proto(int proto_num, const struct pppox_proto *pp);
 extern void unregister_pppox_proto(int proto_num);
