diff --git a/include/linux/hugetlb.h b/include/linux/hugetlb.h
index 48c76d6..5d503f4 100644
--- a/include/linux/hugetlb.h
+++ b/include/linux/hugetlb.h
@@ -314,7 +314,7 @@ struct hstate {
 	unsigned int surplus_huge_pages_node[MAX_NUMNODES];
 #ifdef CONFIG_CGROUP_HUGETLB
 	/* cgroup control files */
-	struct cftype cgroup_files[5];
+	struct cftype (*cgroup_files)[5];
 #endif
 	char name[HSTATE_NAME_LEN];
 };
