diff --git a/include/linux/pid.h b/include/linux/pid.h
index 23705a5..af2bfb4 100644
--- a/include/linux/pid.h
+++ b/include/linux/pid.h
@@ -169,8 +169,8 @@ static inline pid_t pid_nr(struct pid *pid)
 	return nr;
 }
 
-pid_t pid_nr_ns(struct pid *pid, struct pid_namespace *ns);
-pid_t pid_vnr(struct pid *pid);
+pid_t pid_nr_ns(const struct pid *pid, const struct pid_namespace *ns);
+pid_t pid_vnr(const struct pid *pid);
 
 #define do_each_pid_task(pid, type, task)				\
 	do {								\
