diff --git a/include/linux/ctype.h b/include/linux/ctype.h
index f13e4ff..9dc6f74 100644
--- a/include/linux/ctype.h
+++ b/include/linux/ctype.h
@@ -59,7 +59,7 @@ static inline unsigned char __toupper(unsigned char c)
  * Fast implementation of tolower() for internal usage. Do not use in your
  * code.
  */
-static inline char _tolower(const char c)
+static inline unsigned char _tolower(const unsigned char c)
 {
 	return c | 0x20;
 }
