diff --git a/include/net/ipv6.h b/include/net/ipv6.h
index 7f15f95..d381999 100644
--- a/include/net/ipv6.h
+++ b/include/net/ipv6.h
@@ -788,7 +788,7 @@ static inline __be32 ip6_make_flowlabel(struct net *net, struct sk_buff *skb,
 	 * to minimize possbility that any useful information to an
 	 * attacker is leaked. Only lower 20 bits are relevant.
 	 */
-	rol32(hash, 16);
+	hash = rol32(hash, 16);
 
 	flowlabel = (__force __be32)hash & IPV6_FLOWLABEL_MASK;
 
