diff --git a/include/linux/pm_runtime.h b/include/linux/pm_runtime.h
index 2e14d26..aed7c63 100644
--- a/include/linux/pm_runtime.h
+++ b/include/linux/pm_runtime.h
@@ -116,7 +116,7 @@ static inline bool pm_runtime_callbacks_present(struct device *dev)
 
 static inline void pm_runtime_mark_last_busy(struct device *dev)
 {
-	ACCESS_ONCE(dev->power.last_busy) = jiffies;
+	ACCESS_ONCE_RW(dev->power.last_busy) = jiffies;
 }
 
 static inline bool pm_runtime_is_irq_safe(struct device *dev)
