diff --git a/include/linux/rwsem.h b/include/linux/rwsem.h
index dd1d142..4d6d861 100644
--- a/include/linux/rwsem.h
+++ b/include/linux/rwsem.h
@@ -49,6 +49,12 @@ extern struct rw_semaphore *rwsem_down_write_failed_killable(struct rw_semaphore
 extern struct rw_semaphore *rwsem_wake(struct rw_semaphore *);
 extern struct rw_semaphore *rwsem_downgrade_wake(struct rw_semaphore *sem);
 
+extern struct rw_semaphore *call_rwsem_down_read_failed(struct rw_semaphore *sem) __rap_hash;
+extern struct rw_semaphore *call_rwsem_down_write_failed(struct rw_semaphore *sem) __rap_hash;
+extern struct rw_semaphore *call_rwsem_down_write_failed_killable(struct rw_semaphore *sem) __rap_hash;
+extern struct rw_semaphore *call_rwsem_wake(struct rw_semaphore *) __rap_hash;
+extern struct rw_semaphore *call_rwsem_downgrade_wake(struct rw_semaphore *sem) __rap_hash;
+
 /* Include the arch specific part */
 #include <asm/rwsem.h>
 
