diff --git a/include/linux/poison.h b/include/linux/poison.h
index 51334ed..7fda393 100644
--- a/include/linux/poison.h
+++ b/include/linux/poison.h
@@ -19,8 +19,8 @@
  * under normal circumstances, used to verify that nobody uses
  * non-initialized list entries.
  */
-#define LIST_POISON1  ((void *) 0x100 + POISON_POINTER_DELTA)
-#define LIST_POISON2  ((void *) 0x200 + POISON_POINTER_DELTA)
+#define LIST_POISON1  ((void *) (long)0xFFFFFF02)
+#define LIST_POISON2  ((void *) (long)0xFFFFFF04)
 
 /********** include/linux/timer.h **********/
 /*
