diff --git a/include/net/netns/conntrack.h b/include/net/netns/conntrack.h
index e469e85..657e258 100644
--- a/include/net/netns/conntrack.h
+++ b/include/net/netns/conntrack.h
@@ -14,7 +14,7 @@ struct nf_conntrack_ecache;
 struct nf_proto_net {
 #ifdef CONFIG_SYSCTL
 	struct ctl_table_header *ctl_table_header;
-	struct ctl_table        *ctl_table;
+	ctl_table_no_const      *ctl_table;
 #endif
 	unsigned int		users;
 };
