diff --git a/include/drm/drm_mm.h b/include/drm/drm_mm.h
index 205ddcf..0934d31 100644
--- a/include/drm/drm_mm.h
+++ b/include/drm/drm_mm.h
@@ -297,7 +297,7 @@ void drm_mm_remove_node(struct drm_mm_node *node);
 void drm_mm_replace_node(struct drm_mm_node *old, struct drm_mm_node *new);
 void drm_mm_init(struct drm_mm *mm,
 		 u64 start,
-		 u64 size);
+		 u64 size) __intentional_overflow(3);
 void drm_mm_takedown(struct drm_mm *mm);
 bool drm_mm_clean(struct drm_mm *mm);
 
