diff --git a/include/video/uvesafb.h b/include/video/uvesafb.h
index 30f5362..8ed8ac9 100644
--- a/include/video/uvesafb.h
+++ b/include/video/uvesafb.h
@@ -122,6 +122,7 @@ struct uvesafb_par {
 	u8 ypan;			/* 0 - nothing, 1 - ypan, 2 - ywrap */
 	u8 pmi_setpal;			/* PMI for palette changes */
 	u16 *pmi_base;			/* protected mode interface location */
+	u8 *pmi_code;			/* protected mode code location */
 	void *pmi_start;
 	void *pmi_pal;
 	u8 *vbe_state_orig;		/*
