diff --git a/include/net/rtnetlink.h b/include/net/rtnetlink.h
index 4113916..afa5d60 100644
--- a/include/net/rtnetlink.h
+++ b/include/net/rtnetlink.h
@@ -103,7 +103,7 @@ struct rtnl_link_ops {
 	int			(*fill_linkxstats)(struct sk_buff *skb,
 						   const struct net_device *dev,
 						   int *prividx, int attr);
-};
+} __do_const;
 
 int __rtnl_link_register(struct rtnl_link_ops *ops);
 void __rtnl_link_unregister(struct rtnl_link_ops *ops);
