diff --git a/kernel/watchdog.c b/kernel/watchdog.c
index 6d1020c..ae8f990 100644
--- a/kernel/watchdog.c
+++ b/kernel/watchdog.c
@@ -445,7 +445,7 @@ static void watchdog(unsigned int cpu)
 		watchdog_nmi_disable(cpu);
 }
 
-static struct smp_hotplug_thread watchdog_threads = {
+static struct smp_hotplug_thread watchdog_threads __read_only = {
 	.store			= &softlockup_watchdog,
 	.thread_should_run	= watchdog_should_run,
 	.thread_fn		= watchdog,
