diff --git a/kernel/time/timer.c b/kernel/time/timer.c
index c611c47..9133bad 100644
--- a/kernel/time/timer.c
+++ b/kernel/time/timer.c
@@ -1647,7 +1647,7 @@ static inline void __run_timers(struct timer_base *base)
 /*
  * This function runs timers and the timer-tq in bottom half context.
  */
-static __latent_entropy void run_timer_softirq(struct softirq_action *h)
+static __latent_entropy void run_timer_softirq(void)
 {
 	struct timer_base *base = this_cpu_ptr(&timer_bases[BASE_STD]);
 
