diff --git a/kernel/seccomp.c b/kernel/seccomp.c
index 0db7c8a..5e67101 100644
--- a/kernel/seccomp.c
+++ b/kernel/seccomp.c
@@ -654,11 +654,14 @@ static int __seccomp_filter(int this_syscall, const struct seccomp_data *sd,
 }
 #endif
 
-int __secure_computing(const struct seccomp_data *sd)
+int secure_computing(const struct seccomp_data *sd)
 {
 	int mode = current->seccomp.mode;
 	int this_syscall;
 
+	if (likely(!test_thread_flag(TIF_SECCOMP)))
+		return 0;
+
 	if (IS_ENABLED(CONFIG_CHECKPOINT_RESTORE) &&
 	    unlikely(current->ptrace & PT_SUSPEND_SECCOMP))
 		return 0;
