diff --git a/lib/decompress_unlzma.c b/lib/decompress_unlzma.c
index ed7a1fd..44a1a62 100644
--- a/lib/decompress_unlzma.c
+++ b/lib/decompress_unlzma.c
@@ -39,10 +39,10 @@
 
 #define	MIN(a, b) (((a) < (b)) ? (a) : (b))
 
-static long long INIT read_int(unsigned char *ptr, int size)
+static unsigned long long INIT read_int(unsigned char *ptr, int size)
 {
 	int i;
-	long long ret = 0;
+	unsigned long long ret = 0;
 
 	for (i = 0; i < size; i++)
 		ret = (ret << 8) | ptr[size-i-1];
