diff --git a/mm/maccess.c b/mm/maccess.c
index 78f9274..5d8c2e02 100644
--- a/mm/maccess.c
+++ b/mm/maccess.c
@@ -28,12 +28,12 @@ long __probe_kernel_read(void *dst, const void *src, size_t size)
 	long ret;
 	mm_segment_t old_fs = get_fs();
 
-	set_fs(KERNEL_DS);
 	pagefault_disable();
+	set_fs(KERNEL_DS);
 	ret = __copy_from_user_inatomic(dst,
-			(__force const void __user *)src, size);
-	pagefault_enable();
+			(const void __force_user *)src, size);
 	set_fs(old_fs);
+	pagefault_enable();
 
 	return ret ? -EFAULT : 0;
 }
@@ -56,11 +56,11 @@ long __probe_kernel_write(void *dst, const void *src, size_t size)
 	long ret;
 	mm_segment_t old_fs = get_fs();
 
-	set_fs(KERNEL_DS);
 	pagefault_disable();
-	ret = __copy_to_user_inatomic((__force void __user *)dst, src, size);
-	pagefault_enable();
+	set_fs(KERNEL_DS);
+	ret = __copy_to_user_inatomic((void __force_user *)dst, src, size);
 	set_fs(old_fs);
+	pagefault_enable();
 
 	return ret ? -EFAULT : 0;
 }
