diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 47559cc..3c43108 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -701,7 +701,7 @@ static void memcg_check_events(struct mem_cgroup *memcg, struct page *page)
 			mem_cgroup_update_tree(memcg, page);
 #if MAX_NUMNODES > 1
 		if (unlikely(do_numainfo))
-			atomic_inc(&memcg->numainfo_events);
+			atomic64_inc(&memcg->numainfo_events);
 #endif
 	}
 }
@@ -1286,7 +1286,7 @@ static void mem_cgroup_may_update_nodemask(struct mem_cgroup *memcg)
 	 * numainfo_events > 0 means there was at least NUMAINFO_EVENTS_TARGET
 	 * pagein/pageout changes since the last update.
 	 */
-	if (!atomic_read(&memcg->numainfo_events))
+	if (!atomic64_read(&memcg->numainfo_events))
 		return;
 	if (atomic_inc_return(&memcg->numainfo_updating) > 1)
 		return;
@@ -1300,7 +1300,7 @@ static void mem_cgroup_may_update_nodemask(struct mem_cgroup *memcg)
 			node_clear(nid, memcg->scan_nodes);
 	}
 
-	atomic_set(&memcg->numainfo_events, 0);
+	atomic64_set(&memcg->numainfo_events, 0);
 	atomic_set(&memcg->numainfo_updating, 0);
 }
 
