diff --git a/net/rds/tcp.h b/net/rds/tcp.h
index 9a1cc89..d5eb68d 100644
--- a/net/rds/tcp.h
+++ b/net/rds/tcp.h
@@ -17,9 +17,9 @@ struct rds_tcp_connection {
 	 */
 	struct mutex		t_conn_path_lock;
 	struct socket		*t_sock;
-	void			*t_orig_write_space;
-	void			*t_orig_data_ready;
-	void			*t_orig_state_change;
+	void			(*t_orig_write_space)(struct sock *sk);
+	void			(*t_orig_data_ready)(struct sock *sk);
+	void			(*t_orig_state_change)(struct sock *sk);
 
 	struct rds_tcp_incoming	*t_tinc;
 	size_t			t_tinc_hdr_rem;
