diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index e5194f6f..ad76273 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -506,7 +506,7 @@ static int nf_tables_gettable(struct net *net, struct sock *nlsk,
 	int err;
 
 	if (nlh->nlmsg_flags & NLM_F_DUMP) {
-		struct netlink_dump_control c = {
+		static struct netlink_dump_control c = {
 			.dump = nf_tables_dump_tables,
 		};
 		return netlink_dump_start(nlsk, skb, nlh, &c);
@@ -1092,7 +1092,7 @@ static int nf_tables_getchain(struct net *net, struct sock *nlsk,
 	int err;
 
 	if (nlh->nlmsg_flags & NLM_F_DUMP) {
-		struct netlink_dump_control c = {
+		static struct netlink_dump_control c = {
 			.dump = nf_tables_dump_chains,
 		};
 		return netlink_dump_start(nlsk, skb, nlh, &c);
@@ -2001,7 +2001,7 @@ static int nf_tables_getrule(struct net *net, struct sock *nlsk,
 	int err;
 
 	if (nlh->nlmsg_flags & NLM_F_DUMP) {
-		struct netlink_dump_control c = {
+		netlink_dump_control_no_const c = {
 			.dump = nf_tables_dump_rules,
 			.done = nf_tables_dump_rules_done,
 		};
@@ -2724,20 +2724,17 @@ static int nf_tables_getset(struct net *net, struct sock *nlsk,
 		return err;
 
 	if (nlh->nlmsg_flags & NLM_F_DUMP) {
-		struct netlink_dump_control c = {
+		static struct netlink_dump_control c = {
 			.dump = nf_tables_dump_sets,
 			.done = nf_tables_dump_sets_done,
 		};
 		struct nft_ctx *ctx_dump;
 
-		ctx_dump = kmalloc(sizeof(*ctx_dump), GFP_KERNEL);
+		ctx_dump = kmemdup(&ctx, sizeof(ctx), GFP_KERNEL);
 		if (ctx_dump == NULL)
 			return -ENOMEM;
 
-		*ctx_dump = ctx;
-		c.data = ctx_dump;
-
-		return netlink_dump_start(nlsk, skb, nlh, &c);
+		return __netlink_dump_start(nlsk, skb, nlh, &c, ctx_dump, THIS_MODULE);
 	}
 
 	/* Only accept unspec with dump */
@@ -3330,7 +3327,7 @@ static int nf_tables_getsetelem(struct net *net, struct sock *nlsk,
 		return PTR_ERR(set);
 
 	if (nlh->nlmsg_flags & NLM_F_DUMP) {
-		struct netlink_dump_control c = {
+		static struct netlink_dump_control c = {
 			.dump = nf_tables_dump_set,
 		};
 		return netlink_dump_start(nlsk, skb, nlh, &c);
