diff --git a/net/sctp/sysctl.c b/net/sctp/sysctl.c
index daf8554..4ab495b0 100644
--- a/net/sctp/sysctl.c
+++ b/net/sctp/sysctl.c
@@ -324,7 +324,7 @@ static int proc_sctp_do_hmac_alg(struct ctl_table *ctl, int write,
 				loff_t *ppos)
 {
 	struct net *net = current->nsproxy->net_ns;
-	struct ctl_table tbl;
+	ctl_table_no_const tbl;
 	bool changed = false;
 	char *none = "none";
 	char tmp[8] = {0};
@@ -372,7 +372,7 @@ static int proc_sctp_do_rto_min(struct ctl_table *ctl, int write,
 	struct net *net = current->nsproxy->net_ns;
 	unsigned int min = *(unsigned int *) ctl->extra1;
 	unsigned int max = *(unsigned int *) ctl->extra2;
-	struct ctl_table tbl;
+	ctl_table_no_const tbl;
 	int ret, new_value;
 
 	memset(&tbl, 0, sizeof(struct ctl_table));
@@ -401,7 +401,7 @@ static int proc_sctp_do_rto_max(struct ctl_table *ctl, int write,
 	struct net *net = current->nsproxy->net_ns;
 	unsigned int min = *(unsigned int *) ctl->extra1;
 	unsigned int max = *(unsigned int *) ctl->extra2;
-	struct ctl_table tbl;
+	ctl_table_no_const tbl;
 	int ret, new_value;
 
 	memset(&tbl, 0, sizeof(struct ctl_table));
@@ -439,7 +439,7 @@ static int proc_sctp_do_auth(struct ctl_table *ctl, int write,
 			     loff_t *ppos)
 {
 	struct net *net = current->nsproxy->net_ns;
-	struct ctl_table tbl;
+	ctl_table_no_const tbl;
 	int new_value, ret;
 
 	memset(&tbl, 0, sizeof(struct ctl_table));
@@ -466,7 +466,7 @@ static int proc_sctp_do_auth(struct ctl_table *ctl, int write,
 
 int sctp_sysctl_net_register(struct net *net)
 {
-	struct ctl_table *table;
+	ctl_table_no_const *table;
 	int i;
 
 	table = kmemdup(sctp_net_table, sizeof(sctp_net_table), GFP_KERNEL);
