diff --git a/net/ipv6/xfrm6_mode_beet.c b/net/ipv6/xfrm6_mode_beet.c
index 1e205c3..d71b846 100644
--- a/net/ipv6/xfrm6_mode_beet.c
+++ b/net/ipv6/xfrm6_mode_beet.c
@@ -37,7 +37,7 @@ static void xfrm6_beet_make_header(struct sk_buff *skb)
  *
  * The top IP header will be constructed per draft-nikander-esp-beet-mode-06.txt.
  */
-static int xfrm6_beet_output(struct xfrm_state *x, struct sk_buff *skb)
+static int __intentional_overflow(0) xfrm6_beet_output(struct xfrm_state *x, struct sk_buff *skb)
 {
 	struct ipv6hdr *top_iph;
 	struct ip_beet_phdr *ph;
