diff --git a/net/netlink/diag.c b/net/netlink/diag.c
index a554624..640606e 100644
--- a/net/netlink/diag.c
+++ b/net/netlink/diag.c
@@ -208,7 +208,7 @@ static int netlink_diag_handler_dump(struct sk_buff *skb, struct nlmsghdr *h)
 		return -EINVAL;
 
 	if (h->nlmsg_flags & NLM_F_DUMP) {
-		struct netlink_dump_control c = {
+		static struct netlink_dump_control c = {
 			.dump = netlink_diag_dump,
 			.done = netlink_diag_dump_done,
 		};
