diff --git a/net/sunrpc/Kconfig b/net/sunrpc/Kconfig
index 04ce2c0..b1dd1bd 100644
--- a/net/sunrpc/Kconfig
+++ b/net/sunrpc/Kconfig
@@ -38,6 +38,7 @@ config SUNRPC_DEBUG
 	bool "RPC: Enable dprintk debugging"
 	depends on SUNRPC && SYSCTL
 	select DEBUG_FS
+	depends on !GRKERNSEC_KMEM
 	help
 	  This option enables a sysctl-based debugging interface
 	  that is be used by the 'rpcdebug' utility to turn on or off
