diff --git a/net/wireless/scan.c b/net/wireless/scan.c
index 35ad69f..6ac3d49 100644
--- a/net/wireless/scan.c
+++ b/net/wireless/scan.c
@@ -1728,8 +1728,9 @@ static int ieee80211_scan_results(struct cfg80211_registered_device *rdev,
 
 int cfg80211_wext_giwscan(struct net_device *dev,
 			  struct iw_request_info *info,
-			  struct iw_point *data, char *extra)
+			  union iwreq_data *wrqu, char *extra)
 {
+	struct iw_point *data = &wrqu->data;
 	struct cfg80211_registered_device *rdev;
 	int res;
 
