diff --git a/net/netfilter/nf_log.c b/net/netfilter/nf_log.c
index 3dca90d..29b498f 100644
--- a/net/netfilter/nf_log.c
+++ b/net/netfilter/nf_log.c
@@ -413,7 +413,7 @@ static const struct file_operations nflog_file_ops = {
 
 #ifdef CONFIG_SYSCTL
 static char nf_log_sysctl_fnames[NFPROTO_NUMPROTO-NFPROTO_UNSPEC][3];
-static struct ctl_table nf_log_sysctl_table[NFPROTO_NUMPROTO+1];
+static ctl_table_no_const nf_log_sysctl_table[NFPROTO_NUMPROTO+1] __read_only;
 
 static int nf_log_proc_dostring(struct ctl_table *table, int write,
 			 void __user *buffer, size_t *lenp, loff_t *ppos)
@@ -425,7 +425,7 @@ static int nf_log_proc_dostring(struct ctl_table *table, int write,
 	struct net *net = table->extra2;
 
 	if (write) {
-		struct ctl_table tmp = *table;
+		ctl_table_no_const tmp = *table;
 
 		tmp.data = buf;
 		r = proc_dostring(&tmp, write, buffer, lenp, ppos);
@@ -445,13 +445,15 @@ static int nf_log_proc_dostring(struct ctl_table *table, int write,
 		rcu_assign_pointer(net->nf.nf_loggers[tindex], logger);
 		mutex_unlock(&nf_log_mutex);
 	} else {
+		ctl_table_no_const nf_log_table = *table;
+
 		mutex_lock(&nf_log_mutex);
 		logger = nft_log_dereference(net->nf.nf_loggers[tindex]);
 		if (!logger)
-			table->data = "NONE";
+			nf_log_table.data = "NONE";
 		else
-			table->data = logger->name;
-		r = proc_dostring(table, write, buffer, lenp, ppos);
+			nf_log_table.data = logger->name;
+		r = proc_dostring(&nf_log_table, write, buffer, lenp, ppos);
 		mutex_unlock(&nf_log_mutex);
 	}
 
@@ -461,7 +463,7 @@ static int nf_log_proc_dostring(struct ctl_table *table, int write,
 static int netfilter_log_sysctl_init(struct net *net)
 {
 	int i;
-	struct ctl_table *table;
+	ctl_table_no_const *table;
 
 	table = nf_log_sysctl_table;
 	if (!net_eq(net, &init_net)) {
