diff --git a/net/mpls/af_mpls.c b/net/mpls/af_mpls.c
index 1309e2c..6e543c6 100644
--- a/net/mpls/af_mpls.c
+++ b/net/mpls/af_mpls.c
@@ -873,7 +873,7 @@ static int mpls_dev_sysctl_register(struct net_device *dev,
 				    struct mpls_dev *mdev)
 {
 	char path[sizeof("net/mpls/conf/") + IFNAMSIZ];
-	struct ctl_table *table;
+	ctl_table_no_const *table;
 	int i;
 
 	table = kmemdup(&mpls_dev_table, sizeof(mpls_dev_table), GFP_KERNEL);
@@ -957,7 +957,7 @@ static void mpls_ifdown(struct net_device *dev, int event)
 			case NETDEV_CHANGE:
 				nh->nh_flags |= RTNH_F_LINKDOWN;
 				if (event != NETDEV_UNREGISTER)
-					ACCESS_ONCE(rt->rt_nhn_alive) = rt->rt_nhn_alive - 1;
+					ACCESS_ONCE_RW(rt->rt_nhn_alive) = rt->rt_nhn_alive - 1;
 				break;
 			}
 			if (event == NETDEV_UNREGISTER)
@@ -995,7 +995,7 @@ static void mpls_ifup(struct net_device *dev, unsigned int nh_flags)
 			nh->nh_flags &= ~nh_flags;
 		} endfor_nexthops(rt);
 
-		ACCESS_ONCE(rt->rt_nhn_alive) = alive;
+		ACCESS_ONCE_RW(rt->rt_nhn_alive) = alive;
 	}
 }
 
@@ -1622,7 +1622,7 @@ static int mpls_platform_labels(struct ctl_table *table, int write,
 	struct net *net = table->data;
 	int platform_labels = net->mpls.platform_labels;
 	int ret;
-	struct ctl_table tmp = {
+	ctl_table_no_const tmp = {
 		.procname	= table->procname,
 		.data		= &platform_labels,
 		.maxlen		= sizeof(int),
@@ -1652,7 +1652,7 @@ static const struct ctl_table mpls_table[] = {
 
 static int mpls_net_init(struct net *net)
 {
-	struct ctl_table *table;
+	ctl_table_no_const *table;
 
 	net->mpls.platform_labels = 0;
 	net->mpls.platform_label = NULL;
