diff --git a/net/irda/irlap_event.c b/net/irda/irlap_event.c
index 0e1b4d7..7e39b12 100644
--- a/net/irda/irlap_event.c
+++ b/net/irda/irlap_event.c
@@ -163,7 +163,7 @@ static int (*state[])(struct irlap_cb *self, IRLAP_EVENT event,
  *    Poll timer has expired. Normally we must now send a RR frame to the
  *    remote device
  */
-static void irlap_poll_timer_expired(void *data)
+static void irlap_poll_timer_expired(unsigned long data)
 {
 	struct irlap_cb *self = (struct irlap_cb *) data;
 
