diff --git a/net/ipv4/inet_diag.c b/net/ipv4/inet_diag.c
index e4d16fc..369aae5 100644
--- a/net/ipv4/inet_diag.c
+++ b/net/ipv4/inet_diag.c
@@ -1082,7 +1082,7 @@ static int inet_diag_rcv_msg_compat(struct sk_buff *skb, struct nlmsghdr *nlh)
 				return err;
 		}
 		{
-			struct netlink_dump_control c = {
+			static struct netlink_dump_control c = {
 				.dump = inet_diag_dump_compat,
 			};
 			return netlink_dump_start(net->diag_nlsk, skb, nlh, &c);
@@ -1113,7 +1113,7 @@ static int inet_diag_handler_cmd(struct sk_buff *skb, struct nlmsghdr *h)
 				return err;
 		}
 		{
-			struct netlink_dump_control c = {
+			static struct netlink_dump_control c = {
 				.dump = inet_diag_dump,
 			};
 			return netlink_dump_start(net->diag_nlsk, skb, h, &c);
