diff --git a/scripts/headers_install.sh b/scripts/headers_install.sh
index fdebd66..a349e33 100755
--- a/scripts/headers_install.sh
+++ b/scripts/headers_install.sh
@@ -32,6 +32,7 @@ do
 	FILE="$(basename "$i")"
 	sed -r \
 		-e 's/([ \t(])(__user|__force|__iomem)[ \t]/\1/g' \
+		-e 's/__intentional_overflow\([- \t,0-9]*\)//g' \
 		-e 's/__attribute_const__([ \t]|$)/\1/g' \
 		-e 's@^#include <linux/compiler.h>@@' \
 		-e 's/(^|[^a-zA-Z0-9])__packed([^a-zA-Z0-9_]|$)/\1__attribute__((packed))\2/g' \
