diff --git a/scripts/kconfig/symbol.c b/scripts/kconfig/symbol.c
index 2432298..faeeb2e 100644
--- a/scripts/kconfig/symbol.c
+++ b/scripts/kconfig/symbol.c
@@ -970,7 +970,7 @@ const char *sym_escape_string_value(const char *in)
 
 struct sym_match {
 	struct symbol	*sym;
-	off_t		so, eo;
+	regoff_t	so, eo;
 };
 
 /* Compare matched symbols as thus:
@@ -992,8 +992,8 @@ static int sym_rel_comp(const void *sym1, const void *sym2)
 	 * exactly; if this is the case, we can't decide which comes first,
 	 * and we fallback to sorting alphabetically.
 	 */
-	exact1 = (s1->eo - s1->so) == strlen(s1->sym->name);
-	exact2 = (s2->eo - s2->so) == strlen(s2->sym->name);
+	exact1 = (s1->eo - s1->so) == (long)strlen(s1->sym->name);
+	exact2 = (s2->eo - s2->so) == (long)strlen(s2->sym->name);
 	if (exact1 && !exact2)
 		return -1;
 	if (!exact1 && exact2)
