diff --git a/scripts/kconfig/lkc.h b/scripts/kconfig/lkc.h
index 91ca126..5f7cad6 100644
--- a/scripts/kconfig/lkc.h
+++ b/scripts/kconfig/lkc.h
@@ -108,7 +108,8 @@ void menu_add_expr(enum prop_type type, struct expr *expr, struct expr *dep);
 void menu_add_symbol(enum prop_type type, struct symbol *sym, struct expr *dep);
 void menu_add_option(int token, char *arg);
 void menu_finalize(struct menu *parent);
-void menu_set_type(int type);
+enum symbol_type;
+void menu_set_type(enum symbol_type type);
 
 /* util.c */
 struct file *file_lookup(const char *name);
@@ -123,7 +124,7 @@ struct gstr {
 	* when max_width is not zero long lines in string s (if any) get
 	* wrapped not to exceed the max_width value
 	*/
-	int max_width;
+	size_t max_width;
 };
 struct gstr str_new(void);
 void str_free(struct gstr *gs);
