diff --git a/scripts/dtc/livetree.c b/scripts/dtc/livetree.c
index e229b84..7141e8e 100644
--- a/scripts/dtc/livetree.c
+++ b/scripts/dtc/livetree.c
@@ -492,7 +492,7 @@ struct node *get_node_by_phandle(struct node *tree, cell_t phandle)
 {
 	struct node *child, *node;
 
-	assert((phandle != 0) && (phandle != -1));
+	assert((phandle != 0) && (phandle != ~0U));
 
 	if (tree->phandle == phandle) {
 		if (tree->deleted)
@@ -523,7 +523,7 @@ cell_t get_node_phandle(struct node *root, struct node *node)
 {
 	static cell_t phandle = 1; /* FIXME: ick, static local */
 
-	if ((node->phandle != 0) && (node->phandle != -1))
+	if ((node->phandle != 0) && (node->phandle != ~0U))
 		return node->phandle;
 
 	while (get_node_by_phandle(root, phandle))
