diff --git a/scripts/kconfig/menu.c b/scripts/kconfig/menu.c
index aed678e..1a703de 100644
--- a/scripts/kconfig/menu.c
+++ b/scripts/kconfig/menu.c
@@ -109,7 +109,7 @@ void menu_add_dep(struct expr *dep)
 	current_entry->dep = expr_alloc_and(current_entry->dep, menu_check_dep(dep));
 }
 
-void menu_set_type(int type)
+void menu_set_type(enum symbol_type type)
 {
 	struct symbol *sym = current_entry->sym;
 
