diff --git a/security/yama/yama_lsm.c b/security/yama/yama_lsm.c
index 0309f21..0c9bdfb 100644
--- a/security/yama/yama_lsm.c
+++ b/security/yama/yama_lsm.c
@@ -425,7 +425,7 @@ static struct security_hook_list yama_hooks[] = {
 static int yama_dointvec_minmax(struct ctl_table *table, int write,
 				void __user *buffer, size_t *lenp, loff_t *ppos)
 {
-	struct ctl_table table_copy;
+	ctl_table_no_const table_copy;
 
 	if (write && !capable(CAP_SYS_PTRACE))
 		return -EPERM;
