diff --git a/security/keys/keyring.c b/security/keys/keyring.c
index c91e4e0..d3193f6 100644
--- a/security/keys/keyring.c
+++ b/security/keys/keyring.c
@@ -1095,8 +1095,6 @@ static int keyring_detect_cycle(struct key *A, struct key *B)
 int __key_link_begin(struct key *keyring,
 		     const struct keyring_index_key *index_key,
 		     struct assoc_array_edit **_edit)
-	__acquires(&keyring->sem)
-	__acquires(&keyring_serialise_link_sem)
 {
 	struct assoc_array_edit *edit;
 	int ret;
@@ -1196,8 +1194,6 @@ void __key_link(struct key *key, struct assoc_array_edit **_edit)
 void __key_link_end(struct key *keyring,
 		    const struct keyring_index_key *index_key,
 		    struct assoc_array_edit *edit)
-	__releases(&keyring->sem)
-	__releases(&keyring_serialise_link_sem)
 {
 	BUG_ON(index_key->type == NULL);
 	kenter("%d,%s,", keyring->serial, index_key->type->name);
