diff --git a/security/apparmor/include/policy.h b/security/apparmor/include/policy.h
index 52275f0..4c84202 100644
--- a/security/apparmor/include/policy.h
+++ b/security/apparmor/include/policy.h
@@ -134,7 +134,7 @@ struct aa_namespace {
 	struct aa_ns_acct acct;
 	struct aa_profile *unconfined;
 	struct list_head sub_ns;
-	atomic_t uniq_null;
+	atomic_unchecked_t uniq_null;
 	long uniq_id;
 
 	struct dentry *dents[AAFS_NS_SIZEOF];
