diff --git a/sound/pci/aw2/aw2-saa7146.c b/sound/pci/aw2/aw2-saa7146.c
index 1d78904..d9c1056 100644
--- a/sound/pci/aw2/aw2-saa7146.c
+++ b/sound/pci/aw2/aw2-saa7146.c
@@ -262,7 +262,7 @@ void snd_aw2_saa7146_define_it_playback_callback(unsigned int stream_number,
 {
 	if (stream_number < NB_STREAM_PLAYBACK) {
 		arr_substream_it_playback_cb[stream_number].p_it_callback =
-		    (snd_aw2_saa7146_it_cb) p_it_callback;
+		    p_it_callback;
 		arr_substream_it_playback_cb[stream_number].p_callback_param =
 		    (void *)p_callback_param;
 	}
@@ -275,7 +275,7 @@ void snd_aw2_saa7146_define_it_capture_callback(unsigned int stream_number,
 {
 	if (stream_number < NB_STREAM_CAPTURE) {
 		arr_substream_it_capture_cb[stream_number].p_it_callback =
-		    (snd_aw2_saa7146_it_cb) p_it_callback;
+		    p_it_callback;
 		arr_substream_it_capture_cb[stream_number].p_callback_param =
 		    (void *)p_callback_param;
 	}
