diff --git a/sound/pci/ctxfi/ctsrc.c b/sound/pci/ctxfi/ctsrc.c
index a5a72df..f86edb8 100644
--- a/sound/pci/ctxfi/ctsrc.c
+++ b/sound/pci/ctxfi/ctsrc.c
@@ -544,8 +544,9 @@ static int src_mgr_commit_write(struct src_mgr *mgr)
 	return 0;
 }
 
-int src_mgr_create(struct hw *hw, struct src_mgr **rsrc_mgr)
+int src_mgr_create(struct hw *hw, void **_rsrc_mgr)
 {
+	struct src_mgr **rsrc_mgr = (struct src_mgr **)_rsrc_mgr;
 	int err, i;
 	struct src_mgr *src_mgr;
 
@@ -584,8 +585,10 @@ int src_mgr_create(struct hw *hw, struct src_mgr **rsrc_mgr)
 	return err;
 }
 
-int src_mgr_destroy(struct src_mgr *src_mgr)
+int src_mgr_destroy(void *_src_mgr)
 {
+	struct src_mgr *src_mgr = _src_mgr;
+
 	rsc_mgr_uninit(&src_mgr->mgr);
 	kfree(src_mgr);
 
@@ -828,8 +831,9 @@ static int srcimp_imap_delete(struct srcimp_mgr *mgr, struct imapper *entry)
 	return err;
 }
 
-int srcimp_mgr_create(struct hw *hw, struct srcimp_mgr **rsrcimp_mgr)
+int srcimp_mgr_create(struct hw *hw, void **_rsrcimp_mgr)
 {
+	struct srcimp_mgr **rsrcimp_mgr = (struct srcimp_mgr **)_rsrcimp_mgr;
 	int err;
 	struct srcimp_mgr *srcimp_mgr;
 	struct imapper *entry;
@@ -873,8 +877,9 @@ int srcimp_mgr_create(struct hw *hw, struct srcimp_mgr **rsrcimp_mgr)
 	return err;
 }
 
-int srcimp_mgr_destroy(struct srcimp_mgr *srcimp_mgr)
+int srcimp_mgr_destroy(void *_srcimp_mgr)
 {
+	struct srcimp_mgr *srcimp_mgr = _srcimp_mgr;
 	unsigned long flags;
 
 	/* free src input mapper list */
